// @browserstack/wdio-browserstack-service/dist/generated/sdk-messages-observability.js
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.7.7
//   protoc               unknown
// source: sdk-messages-observability.proto
/* eslint-disable */
import { BinaryReader, BinaryWriter } from "@bufbuild/protobuf/wire";
import { TestHub } from "./sdk-messages-testhub.js";
export const protobufPackage = "browserstack.sdk.v1";
function createBaseKeyMessage() {
    return { key: "", message: "" };
}
export const KeyMessage = {
    encode(message, writer = new BinaryWriter()) {
        if (message.key !== "") {
            writer.uint32(10).string(message.key);
        }
        if (message.message !== "") {
            writer.uint32(18).string(message.message);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseKeyMessage();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.key = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.message = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            key: isSet(object.key) ? globalThis.String(object.key) : "",
            message: isSet(object.message) ? globalThis.String(object.message) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.key !== "") {
            obj.key = message.key;
        }
        if (message.message !== "") {
            obj.message = message.message;
        }
        return obj;
    },
};
function createBaseObservability() {
    return {
        success: false,
        options: undefined,
        errorTraceId: undefined,
        code: undefined,
        errors: [],
        optionsAsBytes: undefined,
    };
}
export const Observability = {
    encode(message, writer = new BinaryWriter()) {
        if (message.success !== false) {
            writer.uint32(8).bool(message.success);
        }
        if (message.options !== undefined) {
            Observability_ObservabilityOptions.encode(message.options, writer.uint32(18).fork()).join();
        }
        if (message.errorTraceId !== undefined) {
            writer.uint32(26).string(message.errorTraceId);
        }
        if (message.code !== undefined) {
            writer.uint32(34).string(message.code);
        }
        for (const v of message.errors) {
            KeyMessage.encode(v, writer.uint32(42).fork()).join();
        }
        if (message.optionsAsBytes !== undefined) {
            writer.uint32(50).bytes(message.optionsAsBytes);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseObservability();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.success = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.options = Observability_ObservabilityOptions.decode(reader, reader.uint32());
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.errorTraceId = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.code = reader.string();
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.errors.push(KeyMessage.decode(reader, reader.uint32()));
                    continue;
                }
                case 6: {
                    if (tag !== 50) {
                        break;
                    }
                    message.optionsAsBytes = reader.bytes();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            success: isSet(object.success) ? globalThis.Boolean(object.success) : false,
            options: isSet(object.options) ? Observability_ObservabilityOptions.fromJSON(object.options) : undefined,
            errorTraceId: isSet(object.errorTraceId) ? globalThis.String(object.errorTraceId) : undefined,
            code: isSet(object.code) ? globalThis.String(object.code) : undefined,
            errors: globalThis.Array.isArray(object?.errors) ? object.errors.map((e) => KeyMessage.fromJSON(e)) : [],
            optionsAsBytes: isSet(object.optionsAsBytes) ? bytesFromBase64(object.optionsAsBytes) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.success !== false) {
            obj.success = message.success;
        }
        if (message.options !== undefined) {
            obj.options = Observability_ObservabilityOptions.toJSON(message.options);
        }
        if (message.errorTraceId !== undefined) {
            obj.errorTraceId = message.errorTraceId;
        }
        if (message.code !== undefined) {
            obj.code = message.code;
        }
        if (message.errors?.length) {
            obj.errors = message.errors.map((e) => KeyMessage.toJSON(e));
        }
        if (message.optionsAsBytes !== undefined) {
            obj.optionsAsBytes = base64FromBytes(message.optionsAsBytes);
        }
        return obj;
    },
};
function createBaseObservability_ObservabilityOptions() {
    return { allowScreenshots: undefined };
}
export const Observability_ObservabilityOptions = {
    encode(message, writer = new BinaryWriter()) {
        if (message.allowScreenshots !== undefined) {
            writer.uint32(10).string(message.allowScreenshots);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseObservability_ObservabilityOptions();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.allowScreenshots = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            allowScreenshots: isSet(object.allowScreenshots) ? globalThis.String(object.allowScreenshots) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.allowScreenshots !== undefined) {
            obj.allowScreenshots = message.allowScreenshots;
        }
        return obj;
    },
};
function createBaseObservabilityConfigRequest() {
    return {
        binSessionId: "",
        platformIndex: 0,
        testFrameworkName: "",
        testFrameworkVersion: "",
        hubUrl: undefined,
        capabilities: undefined,
    };
}
export const ObservabilityConfigRequest = {
    encode(message, writer = new BinaryWriter()) {
        if (message.binSessionId !== "") {
            writer.uint32(10).string(message.binSessionId);
        }
        if (message.platformIndex !== 0) {
            writer.uint32(16).uint32(message.platformIndex);
        }
        if (message.testFrameworkName !== "") {
            writer.uint32(26).string(message.testFrameworkName);
        }
        if (message.testFrameworkVersion !== "") {
            writer.uint32(34).string(message.testFrameworkVersion);
        }
        if (message.hubUrl !== undefined) {
            writer.uint32(42).string(message.hubUrl);
        }
        if (message.capabilities !== undefined) {
            writer.uint32(50).string(message.capabilities);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseObservabilityConfigRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.binSessionId = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.platformIndex = reader.uint32();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.testFrameworkName = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.testFrameworkVersion = reader.string();
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.hubUrl = reader.string();
                    continue;
                }
                case 6: {
                    if (tag !== 50) {
                        break;
                    }
                    message.capabilities = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            binSessionId: isSet(object.binSessionId) ? globalThis.String(object.binSessionId) : "",
            platformIndex: isSet(object.platformIndex) ? globalThis.Number(object.platformIndex) : 0,
            testFrameworkName: isSet(object.testFrameworkName) ? globalThis.String(object.testFrameworkName) : "",
            testFrameworkVersion: isSet(object.testFrameworkVersion) ? globalThis.String(object.testFrameworkVersion) : "",
            hubUrl: isSet(object.hubUrl) ? globalThis.String(object.hubUrl) : undefined,
            capabilities: isSet(object.capabilities) ? globalThis.String(object.capabilities) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.binSessionId !== "") {
            obj.binSessionId = message.binSessionId;
        }
        if (message.platformIndex !== 0) {
            obj.platformIndex = Math.round(message.platformIndex);
        }
        if (message.testFrameworkName !== "") {
            obj.testFrameworkName = message.testFrameworkName;
        }
        if (message.testFrameworkVersion !== "") {
            obj.testFrameworkVersion = message.testFrameworkVersion;
        }
        if (message.hubUrl !== undefined) {
            obj.hubUrl = message.hubUrl;
        }
        if (message.capabilities !== undefined) {
            obj.capabilities = message.capabilities;
        }
        return obj;
    },
};
function createBaseObservabilityConfigResponse() {
    return { success: false, testhub: undefined, observability: undefined, capabilities: undefined };
}
export const ObservabilityConfigResponse = {
    encode(message, writer = new BinaryWriter()) {
        if (message.success !== false) {
            writer.uint32(8).bool(message.success);
        }
        if (message.testhub !== undefined) {
            TestHub.encode(message.testhub, writer.uint32(18).fork()).join();
        }
        if (message.observability !== undefined) {
            Observability.encode(message.observability, writer.uint32(26).fork()).join();
        }
        if (message.capabilities !== undefined) {
            writer.uint32(34).string(message.capabilities);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseObservabilityConfigResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.success = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.testhub = TestHub.decode(reader, reader.uint32());
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.observability = Observability.decode(reader, reader.uint32());
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.capabilities = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            success: isSet(object.success) ? globalThis.Boolean(object.success) : false,
            testhub: isSet(object.testhub) ? TestHub.fromJSON(object.testhub) : undefined,
            observability: isSet(object.observability) ? Observability.fromJSON(object.observability) : undefined,
            capabilities: isSet(object.capabilities) ? globalThis.String(object.capabilities) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.success !== false) {
            obj.success = message.success;
        }
        if (message.testhub !== undefined) {
            obj.testhub = TestHub.toJSON(message.testhub);
        }
        if (message.observability !== undefined) {
            obj.observability = Observability.toJSON(message.observability);
        }
        if (message.capabilities !== undefined) {
            obj.capabilities = message.capabilities;
        }
        return obj;
    },
};
function bytesFromBase64(b64) {
    if (globalThis.Buffer) {
        return Uint8Array.from(globalThis.Buffer.from(b64, "base64"));
    }
    else {
        const bin = globalThis.atob(b64);
        const arr = new Uint8Array(bin.length);
        for (let i = 0; i < bin.length; ++i) {
            arr[i] = bin.charCodeAt(i);
        }
        return arr;
    }
}
function base64FromBytes(arr) {
    if (globalThis.Buffer) {
        return globalThis.Buffer.from(arr).toString("base64");
    }
    else {
        const bin = [];
        arr.forEach((byte) => {
            bin.push(globalThis.String.fromCharCode(byte));
        });
        return globalThis.btoa(bin.join(""));
    }
}
function isSet(value) {
    return value !== null && value !== undefined;
}
