// @browserstack/wdio-browserstack-service/dist/generated/sdk-messages-percy.js
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.7.7
//   protoc               unknown
// source: sdk-messages-percy.proto
/* eslint-disable */
import { BinaryReader, BinaryWriter } from "@bufbuild/protobuf/wire";
export const protobufPackage = "browserstack.sdk.v1";
function createBasePercy() {
    return { success: false, percyCaptureMode: "", percyBuildId: "", isPercyAutoEnabled: false };
}
export const Percy = {
    encode(message, writer = new BinaryWriter()) {
        if (message.success !== false) {
            writer.uint32(8).bool(message.success);
        }
        if (message.percyCaptureMode !== "") {
            writer.uint32(18).string(message.percyCaptureMode);
        }
        if (message.percyBuildId !== "") {
            writer.uint32(26).string(message.percyBuildId);
        }
        if (message.isPercyAutoEnabled !== false) {
            writer.uint32(32).bool(message.isPercyAutoEnabled);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePercy();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.success = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.percyCaptureMode = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.percyBuildId = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 32) {
                        break;
                    }
                    message.isPercyAutoEnabled = reader.bool();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            success: isSet(object.success) ? globalThis.Boolean(object.success) : false,
            percyCaptureMode: isSet(object.percyCaptureMode) ? globalThis.String(object.percyCaptureMode) : "",
            percyBuildId: isSet(object.percyBuildId) ? globalThis.String(object.percyBuildId) : "",
            isPercyAutoEnabled: isSet(object.isPercyAutoEnabled) ? globalThis.Boolean(object.isPercyAutoEnabled) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.success !== false) {
            obj.success = message.success;
        }
        if (message.percyCaptureMode !== "") {
            obj.percyCaptureMode = message.percyCaptureMode;
        }
        if (message.percyBuildId !== "") {
            obj.percyBuildId = message.percyBuildId;
        }
        if (message.isPercyAutoEnabled !== false) {
            obj.isPercyAutoEnabled = message.isPercyAutoEnabled;
        }
        return obj;
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
