// @browserstack/wdio-browserstack-service/dist/generated/sdk-messages-accessibility.js
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.7.7
//   protoc               unknown
// source: sdk-messages-accessibility.proto
/* eslint-disable */
import { BinaryReader, BinaryWriter } from "@bufbuild/protobuf/wire";
import { KeyMessage } from "./sdk-messages-observability.js";
import { TestHub } from "./sdk-messages-testhub.js";
export const protobufPackage = "browserstack.sdk.v1";
function createBaseAccessibility() {
    return {
        success: false,
        options: undefined,
        errorTraceId: undefined,
        code: undefined,
        errors: [],
        isAppAccessibility: undefined,
    };
}
export const Accessibility = {
    encode(message, writer = new BinaryWriter()) {
        if (message.success !== false) {
            writer.uint32(8).bool(message.success);
        }
        if (message.options !== undefined) {
            Accessibility_AccessibilityOptions.encode(message.options, writer.uint32(18).fork()).join();
        }
        if (message.errorTraceId !== undefined) {
            writer.uint32(26).string(message.errorTraceId);
        }
        if (message.code !== undefined) {
            writer.uint32(34).string(message.code);
        }
        for (const v of message.errors) {
            KeyMessage.encode(v, writer.uint32(42).fork()).join();
        }
        if (message.isAppAccessibility !== undefined) {
            writer.uint32(48).bool(message.isAppAccessibility);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAccessibility();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.success = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.options = Accessibility_AccessibilityOptions.decode(reader, reader.uint32());
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.errorTraceId = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.code = reader.string();
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.errors.push(KeyMessage.decode(reader, reader.uint32()));
                    continue;
                }
                case 6: {
                    if (tag !== 48) {
                        break;
                    }
                    message.isAppAccessibility = reader.bool();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            success: isSet(object.success) ? globalThis.Boolean(object.success) : false,
            options: isSet(object.options) ? Accessibility_AccessibilityOptions.fromJSON(object.options) : undefined,
            errorTraceId: isSet(object.errorTraceId) ? globalThis.String(object.errorTraceId) : undefined,
            code: isSet(object.code) ? globalThis.String(object.code) : undefined,
            errors: globalThis.Array.isArray(object?.errors) ? object.errors.map((e) => KeyMessage.fromJSON(e)) : [],
            isAppAccessibility: isSet(object.isAppAccessibility) ? globalThis.Boolean(object.isAppAccessibility) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.success !== false) {
            obj.success = message.success;
        }
        if (message.options !== undefined) {
            obj.options = Accessibility_AccessibilityOptions.toJSON(message.options);
        }
        if (message.errorTraceId !== undefined) {
            obj.errorTraceId = message.errorTraceId;
        }
        if (message.code !== undefined) {
            obj.code = message.code;
        }
        if (message.errors?.length) {
            obj.errors = message.errors.map((e) => KeyMessage.toJSON(e));
        }
        if (message.isAppAccessibility !== undefined) {
            obj.isAppAccessibility = message.isAppAccessibility;
        }
        return obj;
    },
};
function createBaseAccessibility_AccessibilityOptions() {
    return { status: undefined, capabilities: [], scripts: [], commandsToWrap: undefined };
}
export const Accessibility_AccessibilityOptions = {
    encode(message, writer = new BinaryWriter()) {
        if (message.status !== undefined) {
            writer.uint32(10).string(message.status);
        }
        for (const v of message.capabilities) {
            Accessibility_AccessibilityOptions_AccessibilityCapability.encode(v, writer.uint32(18).fork()).join();
        }
        for (const v of message.scripts) {
            Accessibility_AccessibilityOptions_AccessibilityScript.encode(v, writer.uint32(26).fork()).join();
        }
        if (message.commandsToWrap !== undefined) {
            Accessibility_AccessibilityOptions_AccessibilityCommandsToWrap.encode(message.commandsToWrap, writer.uint32(34).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAccessibility_AccessibilityOptions();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.status = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.capabilities.push(Accessibility_AccessibilityOptions_AccessibilityCapability.decode(reader, reader.uint32()));
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.scripts.push(Accessibility_AccessibilityOptions_AccessibilityScript.decode(reader, reader.uint32()));
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.commandsToWrap = Accessibility_AccessibilityOptions_AccessibilityCommandsToWrap.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            status: isSet(object.status) ? globalThis.String(object.status) : undefined,
            capabilities: globalThis.Array.isArray(object?.capabilities)
                ? object.capabilities.map((e) => Accessibility_AccessibilityOptions_AccessibilityCapability.fromJSON(e))
                : [],
            scripts: globalThis.Array.isArray(object?.scripts)
                ? object.scripts.map((e) => Accessibility_AccessibilityOptions_AccessibilityScript.fromJSON(e))
                : [],
            commandsToWrap: isSet(object.commandsToWrap)
                ? Accessibility_AccessibilityOptions_AccessibilityCommandsToWrap.fromJSON(object.commandsToWrap)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.status !== undefined) {
            obj.status = message.status;
        }
        if (message.capabilities?.length) {
            obj.capabilities = message.capabilities.map((e) => Accessibility_AccessibilityOptions_AccessibilityCapability.toJSON(e));
        }
        if (message.scripts?.length) {
            obj.scripts = message.scripts.map((e) => Accessibility_AccessibilityOptions_AccessibilityScript.toJSON(e));
        }
        if (message.commandsToWrap !== undefined) {
            obj.commandsToWrap = Accessibility_AccessibilityOptions_AccessibilityCommandsToWrap.toJSON(message.commandsToWrap);
        }
        return obj;
    },
};
function createBaseAccessibility_AccessibilityOptions_AccessibilityCommandsToWrap() {
    return { scriptsToRun: [], commands: [] };
}
export const Accessibility_AccessibilityOptions_AccessibilityCommandsToWrap = {
    encode(message, writer = new BinaryWriter()) {
        for (const v of message.scriptsToRun) {
            writer.uint32(10).string(v);
        }
        for (const v of message.commands) {
            Accessibility_AccessibilityOptions_AccessibilityCommand.encode(v, writer.uint32(18).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAccessibility_AccessibilityOptions_AccessibilityCommandsToWrap();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.scriptsToRun.push(reader.string());
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.commands.push(Accessibility_AccessibilityOptions_AccessibilityCommand.decode(reader, reader.uint32()));
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            scriptsToRun: globalThis.Array.isArray(object?.scriptsToRun)
                ? object.scriptsToRun.map((e) => globalThis.String(e))
                : [],
            commands: globalThis.Array.isArray(object?.commands)
                ? object.commands.map((e) => Accessibility_AccessibilityOptions_AccessibilityCommand.fromJSON(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.scriptsToRun?.length) {
            obj.scriptsToRun = message.scriptsToRun;
        }
        if (message.commands?.length) {
            obj.commands = message.commands.map((e) => Accessibility_AccessibilityOptions_AccessibilityCommand.toJSON(e));
        }
        return obj;
    },
};
function createBaseAccessibility_AccessibilityOptions_AccessibilityCommand() {
    return { name: "", library: "", module: "", class: "", method: "" };
}
export const Accessibility_AccessibilityOptions_AccessibilityCommand = {
    encode(message, writer = new BinaryWriter()) {
        if (message.name !== "") {
            writer.uint32(10).string(message.name);
        }
        if (message.library !== "") {
            writer.uint32(18).string(message.library);
        }
        if (message.module !== "") {
            writer.uint32(26).string(message.module);
        }
        if (message.class !== "") {
            writer.uint32(34).string(message.class);
        }
        if (message.method !== "") {
            writer.uint32(42).string(message.method);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAccessibility_AccessibilityOptions_AccessibilityCommand();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.library = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.module = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.class = reader.string();
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.method = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            name: isSet(object.name) ? globalThis.String(object.name) : "",
            library: isSet(object.library) ? globalThis.String(object.library) : "",
            module: isSet(object.module) ? globalThis.String(object.module) : "",
            class: isSet(object.class) ? globalThis.String(object.class) : "",
            method: isSet(object.method) ? globalThis.String(object.method) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.name !== "") {
            obj.name = message.name;
        }
        if (message.library !== "") {
            obj.library = message.library;
        }
        if (message.module !== "") {
            obj.module = message.module;
        }
        if (message.class !== "") {
            obj.class = message.class;
        }
        if (message.method !== "") {
            obj.method = message.method;
        }
        return obj;
    },
};
function createBaseAccessibility_AccessibilityOptions_AccessibilityScript() {
    return { name: "", command: "" };
}
export const Accessibility_AccessibilityOptions_AccessibilityScript = {
    encode(message, writer = new BinaryWriter()) {
        if (message.name !== "") {
            writer.uint32(10).string(message.name);
        }
        if (message.command !== "") {
            writer.uint32(18).string(message.command);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAccessibility_AccessibilityOptions_AccessibilityScript();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.command = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            name: isSet(object.name) ? globalThis.String(object.name) : "",
            command: isSet(object.command) ? globalThis.String(object.command) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.name !== "") {
            obj.name = message.name;
        }
        if (message.command !== "") {
            obj.command = message.command;
        }
        return obj;
    },
};
function createBaseAccessibility_AccessibilityOptions_AccessibilityCapability() {
    return { name: "", value: "" };
}
export const Accessibility_AccessibilityOptions_AccessibilityCapability = {
    encode(message, writer = new BinaryWriter()) {
        if (message.name !== "") {
            writer.uint32(10).string(message.name);
        }
        if (message.value !== "") {
            writer.uint32(18).string(message.value);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAccessibility_AccessibilityOptions_AccessibilityCapability();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.name = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.value = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            name: isSet(object.name) ? globalThis.String(object.name) : "",
            value: isSet(object.value) ? globalThis.String(object.value) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.name !== "") {
            obj.name = message.name;
        }
        if (message.value !== "") {
            obj.value = message.value;
        }
        return obj;
    },
};
function createBaseAccessibilityConfigRequest() {
    return { binSessionId: "", platformIndex: 0, frameworkName: "", frameworkVersion: "", hubUrl: "" };
}
export const AccessibilityConfigRequest = {
    encode(message, writer = new BinaryWriter()) {
        if (message.binSessionId !== "") {
            writer.uint32(10).string(message.binSessionId);
        }
        if (message.platformIndex !== 0) {
            writer.uint32(16).uint32(message.platformIndex);
        }
        if (message.frameworkName !== "") {
            writer.uint32(26).string(message.frameworkName);
        }
        if (message.frameworkVersion !== "") {
            writer.uint32(34).string(message.frameworkVersion);
        }
        if (message.hubUrl !== "") {
            writer.uint32(42).string(message.hubUrl);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAccessibilityConfigRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.binSessionId = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.platformIndex = reader.uint32();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.frameworkName = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.frameworkVersion = reader.string();
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.hubUrl = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            binSessionId: isSet(object.binSessionId) ? globalThis.String(object.binSessionId) : "",
            platformIndex: isSet(object.platformIndex) ? globalThis.Number(object.platformIndex) : 0,
            frameworkName: isSet(object.frameworkName) ? globalThis.String(object.frameworkName) : "",
            frameworkVersion: isSet(object.frameworkVersion) ? globalThis.String(object.frameworkVersion) : "",
            hubUrl: isSet(object.hubUrl) ? globalThis.String(object.hubUrl) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.binSessionId !== "") {
            obj.binSessionId = message.binSessionId;
        }
        if (message.platformIndex !== 0) {
            obj.platformIndex = Math.round(message.platformIndex);
        }
        if (message.frameworkName !== "") {
            obj.frameworkName = message.frameworkName;
        }
        if (message.frameworkVersion !== "") {
            obj.frameworkVersion = message.frameworkVersion;
        }
        if (message.hubUrl !== "") {
            obj.hubUrl = message.hubUrl;
        }
        return obj;
    },
};
function createBaseAccessibilityConfigResponse() {
    return { success: false, testhub: undefined, accessibility: undefined };
}
export const AccessibilityConfigResponse = {
    encode(message, writer = new BinaryWriter()) {
        if (message.success !== false) {
            writer.uint32(8).bool(message.success);
        }
        if (message.testhub !== undefined) {
            TestHub.encode(message.testhub, writer.uint32(18).fork()).join();
        }
        if (message.accessibility !== undefined) {
            Accessibility.encode(message.accessibility, writer.uint32(26).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAccessibilityConfigResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.success = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.testhub = TestHub.decode(reader, reader.uint32());
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.accessibility = Accessibility.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            success: isSet(object.success) ? globalThis.Boolean(object.success) : false,
            testhub: isSet(object.testhub) ? TestHub.fromJSON(object.testhub) : undefined,
            accessibility: isSet(object.accessibility) ? Accessibility.fromJSON(object.accessibility) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.success !== false) {
            obj.success = message.success;
        }
        if (message.testhub !== undefined) {
            obj.testhub = TestHub.toJSON(message.testhub);
        }
        if (message.accessibility !== undefined) {
            obj.accessibility = Accessibility.toJSON(message.accessibility);
        }
        return obj;
    },
};
function createBaseAccessibilityResultRequest() {
    return { binSessionId: "", testRunId: "", sessionId: "", resultType: "" };
}
export const AccessibilityResultRequest = {
    encode(message, writer = new BinaryWriter()) {
        if (message.binSessionId !== "") {
            writer.uint32(10).string(message.binSessionId);
        }
        if (message.testRunId !== "") {
            writer.uint32(18).string(message.testRunId);
        }
        if (message.sessionId !== "") {
            writer.uint32(26).string(message.sessionId);
        }
        if (message.resultType !== "") {
            writer.uint32(34).string(message.resultType);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAccessibilityResultRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.binSessionId = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.testRunId = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.sessionId = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.resultType = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            binSessionId: isSet(object.binSessionId) ? globalThis.String(object.binSessionId) : "",
            testRunId: isSet(object.testRunId) ? globalThis.String(object.testRunId) : "",
            sessionId: isSet(object.sessionId) ? globalThis.String(object.sessionId) : "",
            resultType: isSet(object.resultType) ? globalThis.String(object.resultType) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.binSessionId !== "") {
            obj.binSessionId = message.binSessionId;
        }
        if (message.testRunId !== "") {
            obj.testRunId = message.testRunId;
        }
        if (message.sessionId !== "") {
            obj.sessionId = message.sessionId;
        }
        if (message.resultType !== "") {
            obj.resultType = message.resultType;
        }
        return obj;
    },
};
function createBaseAccessibilityResultResponse() {
    return { success: false, resultResponse: new Uint8Array(0) };
}
export const AccessibilityResultResponse = {
    encode(message, writer = new BinaryWriter()) {
        if (message.success !== false) {
            writer.uint32(8).bool(message.success);
        }
        if (message.resultResponse.length !== 0) {
            writer.uint32(18).bytes(message.resultResponse);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAccessibilityResultResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.success = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.resultResponse = reader.bytes();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            success: isSet(object.success) ? globalThis.Boolean(object.success) : false,
            resultResponse: isSet(object.resultResponse) ? bytesFromBase64(object.resultResponse) : new Uint8Array(0),
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.success !== false) {
            obj.success = message.success;
        }
        if (message.resultResponse.length !== 0) {
            obj.resultResponse = base64FromBytes(message.resultResponse);
        }
        return obj;
    },
};
function bytesFromBase64(b64) {
    if (globalThis.Buffer) {
        return Uint8Array.from(globalThis.Buffer.from(b64, "base64"));
    }
    else {
        const bin = globalThis.atob(b64);
        const arr = new Uint8Array(bin.length);
        for (let i = 0; i < bin.length; ++i) {
            arr[i] = bin.charCodeAt(i);
        }
        return arr;
    }
}
function base64FromBytes(arr) {
    if (globalThis.Buffer) {
        return globalThis.Buffer.from(arr).toString("base64");
    }
    else {
        const bin = [];
        arr.forEach((byte) => {
            bin.push(globalThis.String.fromCharCode(byte));
        });
        return globalThis.btoa(bin.join(""));
    }
}
function isSet(value) {
    return value !== null && value !== undefined;
}
