// @browserstack/wdio-browserstack-service/dist/generated/sdk-messages.js
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.7.7
//   protoc               unknown
// source: sdk-messages.proto
/* eslint-disable */
import { BinaryReader, BinaryWriter } from "@bufbuild/protobuf/wire";
import { Accessibility } from "./sdk-messages-accessibility.js";
import { Observability } from "./sdk-messages-observability.js";
import { Percy } from "./sdk-messages-percy.js";
import { TestHub } from "./sdk-messages-testhub.js";
export const protobufPackage = "browserstack.sdk.v1";
function createBaseDriverInitRequest() {
    return { binSessionId: "", platformIndex: 0, ref: "", userInputParams: new Uint8Array(0) };
}
export const DriverInitRequest = {
    encode(message, writer = new BinaryWriter()) {
        if (message.binSessionId !== "") {
            writer.uint32(10).string(message.binSessionId);
        }
        if (message.platformIndex !== 0) {
            writer.uint32(16).uint32(message.platformIndex);
        }
        if (message.ref !== "") {
            writer.uint32(26).string(message.ref);
        }
        if (message.userInputParams.length !== 0) {
            writer.uint32(34).bytes(message.userInputParams);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDriverInitRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.binSessionId = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.platformIndex = reader.uint32();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.ref = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.userInputParams = reader.bytes();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            binSessionId: isSet(object.binSessionId) ? globalThis.String(object.binSessionId) : "",
            platformIndex: isSet(object.platformIndex) ? globalThis.Number(object.platformIndex) : 0,
            ref: isSet(object.ref) ? globalThis.String(object.ref) : "",
            userInputParams: isSet(object.userInputParams) ? bytesFromBase64(object.userInputParams) : new Uint8Array(0),
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.binSessionId !== "") {
            obj.binSessionId = message.binSessionId;
        }
        if (message.platformIndex !== 0) {
            obj.platformIndex = Math.round(message.platformIndex);
        }
        if (message.ref !== "") {
            obj.ref = message.ref;
        }
        if (message.userInputParams.length !== 0) {
            obj.userInputParams = base64FromBytes(message.userInputParams);
        }
        return obj;
    },
};
function createBaseDriverInitResponse() {
    return { success: false, binSessionId: "", capabilities: new Uint8Array(0), hubUrl: undefined };
}
export const DriverInitResponse = {
    encode(message, writer = new BinaryWriter()) {
        if (message.success !== false) {
            writer.uint32(8).bool(message.success);
        }
        if (message.binSessionId !== "") {
            writer.uint32(18).string(message.binSessionId);
        }
        if (message.capabilities.length !== 0) {
            writer.uint32(26).bytes(message.capabilities);
        }
        if (message.hubUrl !== undefined) {
            writer.uint32(34).string(message.hubUrl);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDriverInitResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.success = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.binSessionId = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.capabilities = reader.bytes();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.hubUrl = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            success: isSet(object.success) ? globalThis.Boolean(object.success) : false,
            binSessionId: isSet(object.binSessionId) ? globalThis.String(object.binSessionId) : "",
            capabilities: isSet(object.capabilities) ? bytesFromBase64(object.capabilities) : new Uint8Array(0),
            hubUrl: isSet(object.hubUrl) ? globalThis.String(object.hubUrl) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.success !== false) {
            obj.success = message.success;
        }
        if (message.binSessionId !== "") {
            obj.binSessionId = message.binSessionId;
        }
        if (message.capabilities.length !== 0) {
            obj.capabilities = base64FromBytes(message.capabilities);
        }
        if (message.hubUrl !== undefined) {
            obj.hubUrl = message.hubUrl;
        }
        return obj;
    },
};
function createBaseAutomationFrameworkInitRequest() {
    return { binSessionId: "", platformIndex: 0, frameworkName: "", frameworkVersion: "", ref: "", hubUrl: "" };
}
export const AutomationFrameworkInitRequest = {
    encode(message, writer = new BinaryWriter()) {
        if (message.binSessionId !== "") {
            writer.uint32(10).string(message.binSessionId);
        }
        if (message.platformIndex !== 0) {
            writer.uint32(16).uint32(message.platformIndex);
        }
        if (message.frameworkName !== "") {
            writer.uint32(26).string(message.frameworkName);
        }
        if (message.frameworkVersion !== "") {
            writer.uint32(34).string(message.frameworkVersion);
        }
        if (message.ref !== "") {
            writer.uint32(42).string(message.ref);
        }
        if (message.hubUrl !== "") {
            writer.uint32(50).string(message.hubUrl);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAutomationFrameworkInitRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.binSessionId = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.platformIndex = reader.uint32();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.frameworkName = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.frameworkVersion = reader.string();
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.ref = reader.string();
                    continue;
                }
                case 6: {
                    if (tag !== 50) {
                        break;
                    }
                    message.hubUrl = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            binSessionId: isSet(object.binSessionId) ? globalThis.String(object.binSessionId) : "",
            platformIndex: isSet(object.platformIndex) ? globalThis.Number(object.platformIndex) : 0,
            frameworkName: isSet(object.frameworkName) ? globalThis.String(object.frameworkName) : "",
            frameworkVersion: isSet(object.frameworkVersion) ? globalThis.String(object.frameworkVersion) : "",
            ref: isSet(object.ref) ? globalThis.String(object.ref) : "",
            hubUrl: isSet(object.hubUrl) ? globalThis.String(object.hubUrl) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.binSessionId !== "") {
            obj.binSessionId = message.binSessionId;
        }
        if (message.platformIndex !== 0) {
            obj.platformIndex = Math.round(message.platformIndex);
        }
        if (message.frameworkName !== "") {
            obj.frameworkName = message.frameworkName;
        }
        if (message.frameworkVersion !== "") {
            obj.frameworkVersion = message.frameworkVersion;
        }
        if (message.ref !== "") {
            obj.ref = message.ref;
        }
        if (message.hubUrl !== "") {
            obj.hubUrl = message.hubUrl;
        }
        return obj;
    },
};
function createBaseAutomationFrameworkInitResponse() {
    return { success: false, hubUrl: undefined };
}
export const AutomationFrameworkInitResponse = {
    encode(message, writer = new BinaryWriter()) {
        if (message.success !== false) {
            writer.uint32(8).bool(message.success);
        }
        if (message.hubUrl !== undefined) {
            writer.uint32(18).string(message.hubUrl);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAutomationFrameworkInitResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.success = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.hubUrl = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            success: isSet(object.success) ? globalThis.Boolean(object.success) : false,
            hubUrl: isSet(object.hubUrl) ? globalThis.String(object.hubUrl) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.success !== false) {
            obj.success = message.success;
        }
        if (message.hubUrl !== undefined) {
            obj.hubUrl = message.hubUrl;
        }
        return obj;
    },
};
function createBaseAutomationFrameworkStartRequest() {
    return {
        binSessionId: "",
        platformIndex: 0,
        frameworkName: "",
        frameworkVersion: "",
        ref: "",
        frameworkSessionId: "",
        hubUrl: "",
    };
}
export const AutomationFrameworkStartRequest = {
    encode(message, writer = new BinaryWriter()) {
        if (message.binSessionId !== "") {
            writer.uint32(10).string(message.binSessionId);
        }
        if (message.platformIndex !== 0) {
            writer.uint32(16).uint32(message.platformIndex);
        }
        if (message.frameworkName !== "") {
            writer.uint32(26).string(message.frameworkName);
        }
        if (message.frameworkVersion !== "") {
            writer.uint32(34).string(message.frameworkVersion);
        }
        if (message.ref !== "") {
            writer.uint32(42).string(message.ref);
        }
        if (message.frameworkSessionId !== "") {
            writer.uint32(50).string(message.frameworkSessionId);
        }
        if (message.hubUrl !== "") {
            writer.uint32(58).string(message.hubUrl);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAutomationFrameworkStartRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.binSessionId = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.platformIndex = reader.uint32();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.frameworkName = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.frameworkVersion = reader.string();
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.ref = reader.string();
                    continue;
                }
                case 6: {
                    if (tag !== 50) {
                        break;
                    }
                    message.frameworkSessionId = reader.string();
                    continue;
                }
                case 7: {
                    if (tag !== 58) {
                        break;
                    }
                    message.hubUrl = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            binSessionId: isSet(object.binSessionId) ? globalThis.String(object.binSessionId) : "",
            platformIndex: isSet(object.platformIndex) ? globalThis.Number(object.platformIndex) : 0,
            frameworkName: isSet(object.frameworkName) ? globalThis.String(object.frameworkName) : "",
            frameworkVersion: isSet(object.frameworkVersion) ? globalThis.String(object.frameworkVersion) : "",
            ref: isSet(object.ref) ? globalThis.String(object.ref) : "",
            frameworkSessionId: isSet(object.frameworkSessionId) ? globalThis.String(object.frameworkSessionId) : "",
            hubUrl: isSet(object.hubUrl) ? globalThis.String(object.hubUrl) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.binSessionId !== "") {
            obj.binSessionId = message.binSessionId;
        }
        if (message.platformIndex !== 0) {
            obj.platformIndex = Math.round(message.platformIndex);
        }
        if (message.frameworkName !== "") {
            obj.frameworkName = message.frameworkName;
        }
        if (message.frameworkVersion !== "") {
            obj.frameworkVersion = message.frameworkVersion;
        }
        if (message.ref !== "") {
            obj.ref = message.ref;
        }
        if (message.frameworkSessionId !== "") {
            obj.frameworkSessionId = message.frameworkSessionId;
        }
        if (message.hubUrl !== "") {
            obj.hubUrl = message.hubUrl;
        }
        return obj;
    },
};
function createBaseAutomationFrameworkStartResponse() {
    return { success: false };
}
export const AutomationFrameworkStartResponse = {
    encode(message, writer = new BinaryWriter()) {
        if (message.success !== false) {
            writer.uint32(8).bool(message.success);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAutomationFrameworkStartResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.success = reader.bool();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { success: isSet(object.success) ? globalThis.Boolean(object.success) : false };
    },
    toJSON(message) {
        const obj = {};
        if (message.success !== false) {
            obj.success = message.success;
        }
        return obj;
    },
};
function createBaseAutomationFrameworkStopRequest() {
    return {
        binSessionId: "",
        platformIndex: 0,
        frameworkName: "",
        frameworkVersion: "",
        ref: "",
        frameworkSessionId: "",
        hubUrl: "",
    };
}
export const AutomationFrameworkStopRequest = {
    encode(message, writer = new BinaryWriter()) {
        if (message.binSessionId !== "") {
            writer.uint32(10).string(message.binSessionId);
        }
        if (message.platformIndex !== 0) {
            writer.uint32(16).uint32(message.platformIndex);
        }
        if (message.frameworkName !== "") {
            writer.uint32(26).string(message.frameworkName);
        }
        if (message.frameworkVersion !== "") {
            writer.uint32(34).string(message.frameworkVersion);
        }
        if (message.ref !== "") {
            writer.uint32(42).string(message.ref);
        }
        if (message.frameworkSessionId !== "") {
            writer.uint32(50).string(message.frameworkSessionId);
        }
        if (message.hubUrl !== "") {
            writer.uint32(58).string(message.hubUrl);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAutomationFrameworkStopRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.binSessionId = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.platformIndex = reader.uint32();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.frameworkName = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.frameworkVersion = reader.string();
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.ref = reader.string();
                    continue;
                }
                case 6: {
                    if (tag !== 50) {
                        break;
                    }
                    message.frameworkSessionId = reader.string();
                    continue;
                }
                case 7: {
                    if (tag !== 58) {
                        break;
                    }
                    message.hubUrl = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            binSessionId: isSet(object.binSessionId) ? globalThis.String(object.binSessionId) : "",
            platformIndex: isSet(object.platformIndex) ? globalThis.Number(object.platformIndex) : 0,
            frameworkName: isSet(object.frameworkName) ? globalThis.String(object.frameworkName) : "",
            frameworkVersion: isSet(object.frameworkVersion) ? globalThis.String(object.frameworkVersion) : "",
            ref: isSet(object.ref) ? globalThis.String(object.ref) : "",
            frameworkSessionId: isSet(object.frameworkSessionId) ? globalThis.String(object.frameworkSessionId) : "",
            hubUrl: isSet(object.hubUrl) ? globalThis.String(object.hubUrl) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.binSessionId !== "") {
            obj.binSessionId = message.binSessionId;
        }
        if (message.platformIndex !== 0) {
            obj.platformIndex = Math.round(message.platformIndex);
        }
        if (message.frameworkName !== "") {
            obj.frameworkName = message.frameworkName;
        }
        if (message.frameworkVersion !== "") {
            obj.frameworkVersion = message.frameworkVersion;
        }
        if (message.ref !== "") {
            obj.ref = message.ref;
        }
        if (message.frameworkSessionId !== "") {
            obj.frameworkSessionId = message.frameworkSessionId;
        }
        if (message.hubUrl !== "") {
            obj.hubUrl = message.hubUrl;
        }
        return obj;
    },
};
function createBaseAutomationFrameworkStopResponse() {
    return { success: false };
}
export const AutomationFrameworkStopResponse = {
    encode(message, writer = new BinaryWriter()) {
        if (message.success !== false) {
            writer.uint32(8).bool(message.success);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAutomationFrameworkStopResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.success = reader.bool();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { success: isSet(object.success) ? globalThis.Boolean(object.success) : false };
    },
    toJSON(message) {
        const obj = {};
        if (message.success !== false) {
            obj.success = message.success;
        }
        return obj;
    },
};
function createBaseFindNearestHubRequest() {
    return { binSessionId: "" };
}
export const FindNearestHubRequest = {
    encode(message, writer = new BinaryWriter()) {
        if (message.binSessionId !== "") {
            writer.uint32(10).string(message.binSessionId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseFindNearestHubRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.binSessionId = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { binSessionId: isSet(object.binSessionId) ? globalThis.String(object.binSessionId) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.binSessionId !== "") {
            obj.binSessionId = message.binSessionId;
        }
        return obj;
    },
};
function createBaseFindNearestHubResponse() {
    return { url: "", latency: 0 };
}
export const FindNearestHubResponse = {
    encode(message, writer = new BinaryWriter()) {
        if (message.url !== "") {
            writer.uint32(10).string(message.url);
        }
        if (message.latency !== 0) {
            writer.uint32(16).uint32(message.latency);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseFindNearestHubResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.url = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.latency = reader.uint32();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            url: isSet(object.url) ? globalThis.String(object.url) : "",
            latency: isSet(object.latency) ? globalThis.Number(object.latency) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.url !== "") {
            obj.url = message.url;
        }
        if (message.latency !== 0) {
            obj.latency = Math.round(message.latency);
        }
        return obj;
    },
};
function createBaseStartBinSessionRequest() {
    return {
        binSessionId: "",
        sdkLanguage: "",
        sdkVersion: "",
        pathProject: "",
        pathConfig: "",
        envVars: {},
        cliArgs: [],
        frameworks: [],
        frameworkVersions: {},
        language: "",
        languageVersion: undefined,
        testFramework: "",
        wdioConfig: undefined,
    };
}
export const StartBinSessionRequest = {
    encode(message, writer = new BinaryWriter()) {
        if (message.binSessionId !== "") {
            writer.uint32(10).string(message.binSessionId);
        }
        if (message.sdkLanguage !== "") {
            writer.uint32(18).string(message.sdkLanguage);
        }
        if (message.sdkVersion !== "") {
            writer.uint32(26).string(message.sdkVersion);
        }
        if (message.pathProject !== "") {
            writer.uint32(34).string(message.pathProject);
        }
        if (message.pathConfig !== "") {
            writer.uint32(42).string(message.pathConfig);
        }
        Object.entries(message.envVars).forEach(([key, value]) => {
            StartBinSessionRequest_EnvVarsEntry.encode({ key: key, value }, writer.uint32(50).fork()).join();
        });
        for (const v of message.cliArgs) {
            writer.uint32(58).string(v);
        }
        for (const v of message.frameworks) {
            writer.uint32(66).string(v);
        }
        Object.entries(message.frameworkVersions).forEach(([key, value]) => {
            StartBinSessionRequest_FrameworkVersionsEntry.encode({ key: key, value }, writer.uint32(74).fork()).join();
        });
        if (message.language !== "") {
            writer.uint32(82).string(message.language);
        }
        if (message.languageVersion !== undefined) {
            writer.uint32(90).string(message.languageVersion);
        }
        if (message.testFramework !== "") {
            writer.uint32(98).string(message.testFramework);
        }
        if (message.wdioConfig !== undefined) {
            writer.uint32(106).string(message.wdioConfig);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseStartBinSessionRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.binSessionId = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.sdkLanguage = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.sdkVersion = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.pathProject = reader.string();
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.pathConfig = reader.string();
                    continue;
                }
                case 6: {
                    if (tag !== 50) {
                        break;
                    }
                    const entry6 = StartBinSessionRequest_EnvVarsEntry.decode(reader, reader.uint32());
                    if (entry6.value !== undefined) {
                        message.envVars[entry6.key] = entry6.value;
                    }
                    continue;
                }
                case 7: {
                    if (tag !== 58) {
                        break;
                    }
                    message.cliArgs.push(reader.string());
                    continue;
                }
                case 8: {
                    if (tag !== 66) {
                        break;
                    }
                    message.frameworks.push(reader.string());
                    continue;
                }
                case 9: {
                    if (tag !== 74) {
                        break;
                    }
                    const entry9 = StartBinSessionRequest_FrameworkVersionsEntry.decode(reader, reader.uint32());
                    if (entry9.value !== undefined) {
                        message.frameworkVersions[entry9.key] = entry9.value;
                    }
                    continue;
                }
                case 10: {
                    if (tag !== 82) {
                        break;
                    }
                    message.language = reader.string();
                    continue;
                }
                case 11: {
                    if (tag !== 90) {
                        break;
                    }
                    message.languageVersion = reader.string();
                    continue;
                }
                case 12: {
                    if (tag !== 98) {
                        break;
                    }
                    message.testFramework = reader.string();
                    continue;
                }
                case 13: {
                    if (tag !== 106) {
                        break;
                    }
                    message.wdioConfig = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            binSessionId: isSet(object.binSessionId) ? globalThis.String(object.binSessionId) : "",
            sdkLanguage: isSet(object.sdkLanguage) ? globalThis.String(object.sdkLanguage) : "",
            sdkVersion: isSet(object.sdkVersion) ? globalThis.String(object.sdkVersion) : "",
            pathProject: isSet(object.pathProject) ? globalThis.String(object.pathProject) : "",
            pathConfig: isSet(object.pathConfig) ? globalThis.String(object.pathConfig) : "",
            envVars: isObject(object.envVars)
                ? Object.entries(object.envVars).reduce((acc, [key, value]) => {
                    acc[key] = String(value);
                    return acc;
                }, {})
                : {},
            cliArgs: globalThis.Array.isArray(object?.cliArgs) ? object.cliArgs.map((e) => globalThis.String(e)) : [],
            frameworks: globalThis.Array.isArray(object?.frameworks)
                ? object.frameworks.map((e) => globalThis.String(e))
                : [],
            frameworkVersions: isObject(object.frameworkVersions)
                ? Object.entries(object.frameworkVersions).reduce((acc, [key, value]) => {
                    acc[key] = String(value);
                    return acc;
                }, {})
                : {},
            language: isSet(object.language) ? globalThis.String(object.language) : "",
            languageVersion: isSet(object.languageVersion) ? globalThis.String(object.languageVersion) : undefined,
            testFramework: isSet(object.testFramework) ? globalThis.String(object.testFramework) : "",
            wdioConfig: isSet(object.wdioConfig) ? globalThis.String(object.wdioConfig) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.binSessionId !== "") {
            obj.binSessionId = message.binSessionId;
        }
        if (message.sdkLanguage !== "") {
            obj.sdkLanguage = message.sdkLanguage;
        }
        if (message.sdkVersion !== "") {
            obj.sdkVersion = message.sdkVersion;
        }
        if (message.pathProject !== "") {
            obj.pathProject = message.pathProject;
        }
        if (message.pathConfig !== "") {
            obj.pathConfig = message.pathConfig;
        }
        if (message.envVars) {
            const entries = Object.entries(message.envVars);
            if (entries.length > 0) {
                obj.envVars = {};
                entries.forEach(([k, v]) => {
                    obj.envVars[k] = v;
                });
            }
        }
        if (message.cliArgs?.length) {
            obj.cliArgs = message.cliArgs;
        }
        if (message.frameworks?.length) {
            obj.frameworks = message.frameworks;
        }
        if (message.frameworkVersions) {
            const entries = Object.entries(message.frameworkVersions);
            if (entries.length > 0) {
                obj.frameworkVersions = {};
                entries.forEach(([k, v]) => {
                    obj.frameworkVersions[k] = v;
                });
            }
        }
        if (message.language !== "") {
            obj.language = message.language;
        }
        if (message.languageVersion !== undefined) {
            obj.languageVersion = message.languageVersion;
        }
        if (message.testFramework !== "") {
            obj.testFramework = message.testFramework;
        }
        if (message.wdioConfig !== undefined) {
            obj.wdioConfig = message.wdioConfig;
        }
        return obj;
    },
};
function createBaseStartBinSessionRequest_EnvVarsEntry() {
    return { key: "", value: "" };
}
export const StartBinSessionRequest_EnvVarsEntry = {
    encode(message, writer = new BinaryWriter()) {
        if (message.key !== "") {
            writer.uint32(10).string(message.key);
        }
        if (message.value !== "") {
            writer.uint32(18).string(message.value);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseStartBinSessionRequest_EnvVarsEntry();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.key = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.value = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            key: isSet(object.key) ? globalThis.String(object.key) : "",
            value: isSet(object.value) ? globalThis.String(object.value) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.key !== "") {
            obj.key = message.key;
        }
        if (message.value !== "") {
            obj.value = message.value;
        }
        return obj;
    },
};
function createBaseStartBinSessionRequest_FrameworkVersionsEntry() {
    return { key: "", value: "" };
}
export const StartBinSessionRequest_FrameworkVersionsEntry = {
    encode(message, writer = new BinaryWriter()) {
        if (message.key !== "") {
            writer.uint32(10).string(message.key);
        }
        if (message.value !== "") {
            writer.uint32(18).string(message.value);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseStartBinSessionRequest_FrameworkVersionsEntry();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.key = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.value = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            key: isSet(object.key) ? globalThis.String(object.key) : "",
            value: isSet(object.value) ? globalThis.String(object.value) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.key !== "") {
            obj.key = message.key;
        }
        if (message.value !== "") {
            obj.value = message.value;
        }
        return obj;
    },
};
function createBaseStartBinSessionResponse() {
    return {
        binSessionId: "",
        config: "",
        testhub: undefined,
        observability: undefined,
        accessibility: undefined,
        percy: undefined,
        sessionFramework: undefined,
    };
}
export const StartBinSessionResponse = {
    encode(message, writer = new BinaryWriter()) {
        if (message.binSessionId !== "") {
            writer.uint32(10).string(message.binSessionId);
        }
        if (message.config !== "") {
            writer.uint32(18).string(message.config);
        }
        if (message.testhub !== undefined) {
            TestHub.encode(message.testhub, writer.uint32(26).fork()).join();
        }
        if (message.observability !== undefined) {
            Observability.encode(message.observability, writer.uint32(34).fork()).join();
        }
        if (message.accessibility !== undefined) {
            Accessibility.encode(message.accessibility, writer.uint32(42).fork()).join();
        }
        if (message.percy !== undefined) {
            Percy.encode(message.percy, writer.uint32(50).fork()).join();
        }
        if (message.sessionFramework !== undefined) {
            writer.uint32(58).string(message.sessionFramework);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseStartBinSessionResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.binSessionId = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.config = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.testhub = TestHub.decode(reader, reader.uint32());
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.observability = Observability.decode(reader, reader.uint32());
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.accessibility = Accessibility.decode(reader, reader.uint32());
                    continue;
                }
                case 6: {
                    if (tag !== 50) {
                        break;
                    }
                    message.percy = Percy.decode(reader, reader.uint32());
                    continue;
                }
                case 7: {
                    if (tag !== 58) {
                        break;
                    }
                    message.sessionFramework = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            binSessionId: isSet(object.binSessionId) ? globalThis.String(object.binSessionId) : "",
            config: isSet(object.config) ? globalThis.String(object.config) : "",
            testhub: isSet(object.testhub) ? TestHub.fromJSON(object.testhub) : undefined,
            observability: isSet(object.observability) ? Observability.fromJSON(object.observability) : undefined,
            accessibility: isSet(object.accessibility) ? Accessibility.fromJSON(object.accessibility) : undefined,
            percy: isSet(object.percy) ? Percy.fromJSON(object.percy) : undefined,
            sessionFramework: isSet(object.sessionFramework) ? globalThis.String(object.sessionFramework) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.binSessionId !== "") {
            obj.binSessionId = message.binSessionId;
        }
        if (message.config !== "") {
            obj.config = message.config;
        }
        if (message.testhub !== undefined) {
            obj.testhub = TestHub.toJSON(message.testhub);
        }
        if (message.observability !== undefined) {
            obj.observability = Observability.toJSON(message.observability);
        }
        if (message.accessibility !== undefined) {
            obj.accessibility = Accessibility.toJSON(message.accessibility);
        }
        if (message.percy !== undefined) {
            obj.percy = Percy.toJSON(message.percy);
        }
        if (message.sessionFramework !== undefined) {
            obj.sessionFramework = message.sessionFramework;
        }
        return obj;
    },
};
function createBaseConnectBinSessionResponse() {
    return {
        binSessionId: "",
        config: "",
        testhub: undefined,
        observability: undefined,
        accessibility: undefined,
        percy: undefined,
        sessionFramework: undefined,
        testOrchestration: undefined,
    };
}
export const ConnectBinSessionResponse = {
    encode(message, writer = new BinaryWriter()) {
        if (message.binSessionId !== "") {
            writer.uint32(10).string(message.binSessionId);
        }
        if (message.config !== "") {
            writer.uint32(18).string(message.config);
        }
        if (message.testhub !== undefined) {
            TestHub.encode(message.testhub, writer.uint32(26).fork()).join();
        }
        if (message.observability !== undefined) {
            Observability.encode(message.observability, writer.uint32(34).fork()).join();
        }
        if (message.accessibility !== undefined) {
            Accessibility.encode(message.accessibility, writer.uint32(42).fork()).join();
        }
        if (message.percy !== undefined) {
            Percy.encode(message.percy, writer.uint32(50).fork()).join();
        }
        if (message.sessionFramework !== undefined) {
            writer.uint32(58).string(message.sessionFramework);
        }
        if (message.testOrchestration !== undefined) {
            TestOrchestration.encode(message.testOrchestration, writer.uint32(66).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseConnectBinSessionResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.binSessionId = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.config = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.testhub = TestHub.decode(reader, reader.uint32());
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.observability = Observability.decode(reader, reader.uint32());
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.accessibility = Accessibility.decode(reader, reader.uint32());
                    continue;
                }
                case 6: {
                    if (tag !== 50) {
                        break;
                    }
                    message.percy = Percy.decode(reader, reader.uint32());
                    continue;
                }
                case 7: {
                    if (tag !== 58) {
                        break;
                    }
                    message.sessionFramework = reader.string();
                    continue;
                }
                case 8: {
                    if (tag !== 66) {
                        break;
                    }
                    message.testOrchestration = TestOrchestration.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            binSessionId: isSet(object.binSessionId) ? globalThis.String(object.binSessionId) : "",
            config: isSet(object.config) ? globalThis.String(object.config) : "",
            testhub: isSet(object.testhub) ? TestHub.fromJSON(object.testhub) : undefined,
            observability: isSet(object.observability) ? Observability.fromJSON(object.observability) : undefined,
            accessibility: isSet(object.accessibility) ? Accessibility.fromJSON(object.accessibility) : undefined,
            percy: isSet(object.percy) ? Percy.fromJSON(object.percy) : undefined,
            sessionFramework: isSet(object.sessionFramework) ? globalThis.String(object.sessionFramework) : undefined,
            testOrchestration: isSet(object.testOrchestration)
                ? TestOrchestration.fromJSON(object.testOrchestration)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.binSessionId !== "") {
            obj.binSessionId = message.binSessionId;
        }
        if (message.config !== "") {
            obj.config = message.config;
        }
        if (message.testhub !== undefined) {
            obj.testhub = TestHub.toJSON(message.testhub);
        }
        if (message.observability !== undefined) {
            obj.observability = Observability.toJSON(message.observability);
        }
        if (message.accessibility !== undefined) {
            obj.accessibility = Accessibility.toJSON(message.accessibility);
        }
        if (message.percy !== undefined) {
            obj.percy = Percy.toJSON(message.percy);
        }
        if (message.sessionFramework !== undefined) {
            obj.sessionFramework = message.sessionFramework;
        }
        if (message.testOrchestration !== undefined) {
            obj.testOrchestration = TestOrchestration.toJSON(message.testOrchestration);
        }
        return obj;
    },
};
function createBaseTestOrchestration() {
    return {
        runPreviouslyFailedFirst: undefined,
        retryTestsOnFailure: undefined,
        abortBuildOnFailure: undefined,
        runSmartSelection: undefined,
    };
}
export const TestOrchestration = {
    encode(message, writer = new BinaryWriter()) {
        if (message.runPreviouslyFailedFirst !== undefined) {
            writer.uint32(8).bool(message.runPreviouslyFailedFirst);
        }
        if (message.retryTestsOnFailure !== undefined) {
            RetryTestsOnFailure.encode(message.retryTestsOnFailure, writer.uint32(18).fork()).join();
        }
        if (message.abortBuildOnFailure !== undefined) {
            AbortBuildOnFailure.encode(message.abortBuildOnFailure, writer.uint32(26).fork()).join();
        }
        if (message.runSmartSelection !== undefined) {
            RunSmartSelection.encode(message.runSmartSelection, writer.uint32(34).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTestOrchestration();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.runPreviouslyFailedFirst = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.retryTestsOnFailure = RetryTestsOnFailure.decode(reader, reader.uint32());
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.abortBuildOnFailure = AbortBuildOnFailure.decode(reader, reader.uint32());
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.runSmartSelection = RunSmartSelection.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            runPreviouslyFailedFirst: isSet(object.runPreviouslyFailedFirst)
                ? globalThis.Boolean(object.runPreviouslyFailedFirst)
                : undefined,
            retryTestsOnFailure: isSet(object.retryTestsOnFailure)
                ? RetryTestsOnFailure.fromJSON(object.retryTestsOnFailure)
                : undefined,
            abortBuildOnFailure: isSet(object.abortBuildOnFailure)
                ? AbortBuildOnFailure.fromJSON(object.abortBuildOnFailure)
                : undefined,
            runSmartSelection: isSet(object.runSmartSelection)
                ? RunSmartSelection.fromJSON(object.runSmartSelection)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.runPreviouslyFailedFirst !== undefined) {
            obj.runPreviouslyFailedFirst = message.runPreviouslyFailedFirst;
        }
        if (message.retryTestsOnFailure !== undefined) {
            obj.retryTestsOnFailure = RetryTestsOnFailure.toJSON(message.retryTestsOnFailure);
        }
        if (message.abortBuildOnFailure !== undefined) {
            obj.abortBuildOnFailure = AbortBuildOnFailure.toJSON(message.abortBuildOnFailure);
        }
        if (message.runSmartSelection !== undefined) {
            obj.runSmartSelection = RunSmartSelection.toJSON(message.runSmartSelection);
        }
        return obj;
    },
};
function createBaseRunSmartSelection() {
    return { enabled: false, mode: "", source: [] };
}
export const RunSmartSelection = {
    encode(message, writer = new BinaryWriter()) {
        if (message.enabled !== false) {
            writer.uint32(8).bool(message.enabled);
        }
        if (message.mode !== "") {
            writer.uint32(18).string(message.mode);
        }
        for (const v of message.source) {
            writer.uint32(26).string(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRunSmartSelection();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.enabled = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.mode = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.source.push(reader.string());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            enabled: isSet(object.enabled) ? globalThis.Boolean(object.enabled) : false,
            mode: isSet(object.mode) ? globalThis.String(object.mode) : "",
            source: globalThis.Array.isArray(object?.source) ? object.source.map((e) => globalThis.String(e)) : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.enabled !== false) {
            obj.enabled = message.enabled;
        }
        if (message.mode !== "") {
            obj.mode = message.mode;
        }
        if (message.source?.length) {
            obj.source = message.source;
        }
        return obj;
    },
};
function createBaseRetryTestsOnFailure() {
    return { enabled: false, maxRetries: 0 };
}
export const RetryTestsOnFailure = {
    encode(message, writer = new BinaryWriter()) {
        if (message.enabled !== false) {
            writer.uint32(8).bool(message.enabled);
        }
        if (message.maxRetries !== 0) {
            writer.uint32(16).uint32(message.maxRetries);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRetryTestsOnFailure();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.enabled = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.maxRetries = reader.uint32();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            enabled: isSet(object.enabled) ? globalThis.Boolean(object.enabled) : false,
            maxRetries: isSet(object.maxRetries) ? globalThis.Number(object.maxRetries) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.enabled !== false) {
            obj.enabled = message.enabled;
        }
        if (message.maxRetries !== 0) {
            obj.maxRetries = Math.round(message.maxRetries);
        }
        return obj;
    },
};
function createBaseAbortBuildOnFailure() {
    return { enabled: false, maxFailures: 0 };
}
export const AbortBuildOnFailure = {
    encode(message, writer = new BinaryWriter()) {
        if (message.enabled !== false) {
            writer.uint32(8).bool(message.enabled);
        }
        if (message.maxFailures !== 0) {
            writer.uint32(16).uint32(message.maxFailures);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAbortBuildOnFailure();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.enabled = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.maxFailures = reader.uint32();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            enabled: isSet(object.enabled) ? globalThis.Boolean(object.enabled) : false,
            maxFailures: isSet(object.maxFailures) ? globalThis.Number(object.maxFailures) : 0,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.enabled !== false) {
            obj.enabled = message.enabled;
        }
        if (message.maxFailures !== 0) {
            obj.maxFailures = Math.round(message.maxFailures);
        }
        return obj;
    },
};
function createBaseTestOrchestrationRequest() {
    return { binSessionId: "", testFiles: [], orchestrationStrategy: undefined, orchestrationMetadata: undefined };
}
export const TestOrchestrationRequest = {
    encode(message, writer = new BinaryWriter()) {
        if (message.binSessionId !== "") {
            writer.uint32(10).string(message.binSessionId);
        }
        for (const v of message.testFiles) {
            writer.uint32(18).string(v);
        }
        if (message.orchestrationStrategy !== undefined) {
            writer.uint32(26).string(message.orchestrationStrategy);
        }
        if (message.orchestrationMetadata !== undefined) {
            writer.uint32(34).string(message.orchestrationMetadata);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTestOrchestrationRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.binSessionId = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.testFiles.push(reader.string());
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.orchestrationStrategy = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.orchestrationMetadata = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            binSessionId: isSet(object.binSessionId) ? globalThis.String(object.binSessionId) : "",
            testFiles: globalThis.Array.isArray(object?.testFiles)
                ? object.testFiles.map((e) => globalThis.String(e))
                : [],
            orchestrationStrategy: isSet(object.orchestrationStrategy)
                ? globalThis.String(object.orchestrationStrategy)
                : undefined,
            orchestrationMetadata: isSet(object.orchestrationMetadata)
                ? globalThis.String(object.orchestrationMetadata)
                : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.binSessionId !== "") {
            obj.binSessionId = message.binSessionId;
        }
        if (message.testFiles?.length) {
            obj.testFiles = message.testFiles;
        }
        if (message.orchestrationStrategy !== undefined) {
            obj.orchestrationStrategy = message.orchestrationStrategy;
        }
        if (message.orchestrationMetadata !== undefined) {
            obj.orchestrationMetadata = message.orchestrationMetadata;
        }
        return obj;
    },
};
function createBaseTestOrchestrationResponse() {
    return { success: false, orderedTestFiles: [] };
}
export const TestOrchestrationResponse = {
    encode(message, writer = new BinaryWriter()) {
        if (message.success !== false) {
            writer.uint32(8).bool(message.success);
        }
        for (const v of message.orderedTestFiles) {
            writer.uint32(18).string(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTestOrchestrationResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.success = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.orderedTestFiles.push(reader.string());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            success: isSet(object.success) ? globalThis.Boolean(object.success) : false,
            orderedTestFiles: globalThis.Array.isArray(object?.orderedTestFiles)
                ? object.orderedTestFiles.map((e) => globalThis.String(e))
                : [],
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.success !== false) {
            obj.success = message.success;
        }
        if (message.orderedTestFiles?.length) {
            obj.orderedTestFiles = message.orderedTestFiles;
        }
        return obj;
    },
};
function createBaseStopBinSessionRequest() {
    return {
        binSessionId: "",
        exitCode: undefined,
        exitSignal: undefined,
        exitReason: undefined,
        customMetadata: undefined,
    };
}
export const StopBinSessionRequest = {
    encode(message, writer = new BinaryWriter()) {
        if (message.binSessionId !== "") {
            writer.uint32(10).string(message.binSessionId);
        }
        if (message.exitCode !== undefined) {
            writer.uint32(16).uint32(message.exitCode);
        }
        if (message.exitSignal !== undefined) {
            writer.uint32(26).string(message.exitSignal);
        }
        if (message.exitReason !== undefined) {
            writer.uint32(34).string(message.exitReason);
        }
        if (message.customMetadata !== undefined) {
            writer.uint32(42).string(message.customMetadata);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseStopBinSessionRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.binSessionId = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.exitCode = reader.uint32();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.exitSignal = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.exitReason = reader.string();
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.customMetadata = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            binSessionId: isSet(object.binSessionId) ? globalThis.String(object.binSessionId) : "",
            exitCode: isSet(object.exitCode) ? globalThis.Number(object.exitCode) : undefined,
            exitSignal: isSet(object.exitSignal) ? globalThis.String(object.exitSignal) : undefined,
            exitReason: isSet(object.exitReason) ? globalThis.String(object.exitReason) : undefined,
            customMetadata: isSet(object.customMetadata) ? globalThis.String(object.customMetadata) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.binSessionId !== "") {
            obj.binSessionId = message.binSessionId;
        }
        if (message.exitCode !== undefined) {
            obj.exitCode = Math.round(message.exitCode);
        }
        if (message.exitSignal !== undefined) {
            obj.exitSignal = message.exitSignal;
        }
        if (message.exitReason !== undefined) {
            obj.exitReason = message.exitReason;
        }
        if (message.customMetadata !== undefined) {
            obj.customMetadata = message.customMetadata;
        }
        return obj;
    },
};
function createBaseStopBinSessionResponse() {
    return { success: false, error: undefined, automateBuildlink: undefined, hashedId: undefined };
}
export const StopBinSessionResponse = {
    encode(message, writer = new BinaryWriter()) {
        if (message.success !== false) {
            writer.uint32(8).bool(message.success);
        }
        if (message.error !== undefined) {
            writer.uint32(18).string(message.error);
        }
        if (message.automateBuildlink !== undefined) {
            writer.uint32(26).string(message.automateBuildlink);
        }
        if (message.hashedId !== undefined) {
            writer.uint32(34).string(message.hashedId);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseStopBinSessionResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.success = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.error = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.automateBuildlink = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.hashedId = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            success: isSet(object.success) ? globalThis.Boolean(object.success) : false,
            error: isSet(object.error) ? globalThis.String(object.error) : undefined,
            automateBuildlink: isSet(object.automateBuildlink) ? globalThis.String(object.automateBuildlink) : undefined,
            hashedId: isSet(object.hashedId) ? globalThis.String(object.hashedId) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.success !== false) {
            obj.success = message.success;
        }
        if (message.error !== undefined) {
            obj.error = message.error;
        }
        if (message.automateBuildlink !== undefined) {
            obj.automateBuildlink = message.automateBuildlink;
        }
        if (message.hashedId !== undefined) {
            obj.hashedId = message.hashedId;
        }
        return obj;
    },
};
function createBaseConnectBinSessionRequest() {
    return { binSessionId: "", platformIndex: undefined };
}
export const ConnectBinSessionRequest = {
    encode(message, writer = new BinaryWriter()) {
        if (message.binSessionId !== "") {
            writer.uint32(10).string(message.binSessionId);
        }
        if (message.platformIndex !== undefined) {
            writer.uint32(16).uint32(message.platformIndex);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseConnectBinSessionRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.binSessionId = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.platformIndex = reader.uint32();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            binSessionId: isSet(object.binSessionId) ? globalThis.String(object.binSessionId) : "",
            platformIndex: isSet(object.platformIndex) ? globalThis.Number(object.platformIndex) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.binSessionId !== "") {
            obj.binSessionId = message.binSessionId;
        }
        if (message.platformIndex !== undefined) {
            obj.platformIndex = Math.round(message.platformIndex);
        }
        return obj;
    },
};
function createBaseTestFrameworkEventRequest() {
    return {
        binSessionId: "",
        platformIndex: 0,
        testFrameworkName: "",
        testFrameworkVersion: "",
        uuid: "",
        eventJson: new Uint8Array(0),
        testFrameworkState: "",
        testHookState: "",
        startedAt: "",
        endedAt: undefined,
        executionContext: undefined,
    };
}
export const TestFrameworkEventRequest = {
    encode(message, writer = new BinaryWriter()) {
        if (message.binSessionId !== "") {
            writer.uint32(10).string(message.binSessionId);
        }
        if (message.platformIndex !== 0) {
            writer.uint32(16).uint32(message.platformIndex);
        }
        if (message.testFrameworkName !== "") {
            writer.uint32(26).string(message.testFrameworkName);
        }
        if (message.testFrameworkVersion !== "") {
            writer.uint32(34).string(message.testFrameworkVersion);
        }
        if (message.uuid !== "") {
            writer.uint32(42).string(message.uuid);
        }
        if (message.eventJson.length !== 0) {
            writer.uint32(50).bytes(message.eventJson);
        }
        if (message.testFrameworkState !== "") {
            writer.uint32(58).string(message.testFrameworkState);
        }
        if (message.testHookState !== "") {
            writer.uint32(66).string(message.testHookState);
        }
        if (message.startedAt !== "") {
            writer.uint32(74).string(message.startedAt);
        }
        if (message.endedAt !== undefined) {
            writer.uint32(82).string(message.endedAt);
        }
        if (message.executionContext !== undefined) {
            ExecutionContext.encode(message.executionContext, writer.uint32(90).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTestFrameworkEventRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.binSessionId = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.platformIndex = reader.uint32();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.testFrameworkName = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.testFrameworkVersion = reader.string();
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.uuid = reader.string();
                    continue;
                }
                case 6: {
                    if (tag !== 50) {
                        break;
                    }
                    message.eventJson = reader.bytes();
                    continue;
                }
                case 7: {
                    if (tag !== 58) {
                        break;
                    }
                    message.testFrameworkState = reader.string();
                    continue;
                }
                case 8: {
                    if (tag !== 66) {
                        break;
                    }
                    message.testHookState = reader.string();
                    continue;
                }
                case 9: {
                    if (tag !== 74) {
                        break;
                    }
                    message.startedAt = reader.string();
                    continue;
                }
                case 10: {
                    if (tag !== 82) {
                        break;
                    }
                    message.endedAt = reader.string();
                    continue;
                }
                case 11: {
                    if (tag !== 90) {
                        break;
                    }
                    message.executionContext = ExecutionContext.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            binSessionId: isSet(object.binSessionId) ? globalThis.String(object.binSessionId) : "",
            platformIndex: isSet(object.platformIndex) ? globalThis.Number(object.platformIndex) : 0,
            testFrameworkName: isSet(object.testFrameworkName) ? globalThis.String(object.testFrameworkName) : "",
            testFrameworkVersion: isSet(object.testFrameworkVersion) ? globalThis.String(object.testFrameworkVersion) : "",
            uuid: isSet(object.uuid) ? globalThis.String(object.uuid) : "",
            eventJson: isSet(object.eventJson) ? bytesFromBase64(object.eventJson) : new Uint8Array(0),
            testFrameworkState: isSet(object.testFrameworkState) ? globalThis.String(object.testFrameworkState) : "",
            testHookState: isSet(object.testHookState) ? globalThis.String(object.testHookState) : "",
            startedAt: isSet(object.startedAt) ? globalThis.String(object.startedAt) : "",
            endedAt: isSet(object.endedAt) ? globalThis.String(object.endedAt) : undefined,
            executionContext: isSet(object.executionContext) ? ExecutionContext.fromJSON(object.executionContext) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.binSessionId !== "") {
            obj.binSessionId = message.binSessionId;
        }
        if (message.platformIndex !== 0) {
            obj.platformIndex = Math.round(message.platformIndex);
        }
        if (message.testFrameworkName !== "") {
            obj.testFrameworkName = message.testFrameworkName;
        }
        if (message.testFrameworkVersion !== "") {
            obj.testFrameworkVersion = message.testFrameworkVersion;
        }
        if (message.uuid !== "") {
            obj.uuid = message.uuid;
        }
        if (message.eventJson.length !== 0) {
            obj.eventJson = base64FromBytes(message.eventJson);
        }
        if (message.testFrameworkState !== "") {
            obj.testFrameworkState = message.testFrameworkState;
        }
        if (message.testHookState !== "") {
            obj.testHookState = message.testHookState;
        }
        if (message.startedAt !== "") {
            obj.startedAt = message.startedAt;
        }
        if (message.endedAt !== undefined) {
            obj.endedAt = message.endedAt;
        }
        if (message.executionContext !== undefined) {
            obj.executionContext = ExecutionContext.toJSON(message.executionContext);
        }
        return obj;
    },
};
function createBaseExecutionContext() {
    return { threadId: "", processId: undefined, hash: undefined };
}
export const ExecutionContext = {
    encode(message, writer = new BinaryWriter()) {
        if (message.threadId !== "") {
            writer.uint32(10).string(message.threadId);
        }
        if (message.processId !== undefined) {
            writer.uint32(18).string(message.processId);
        }
        if (message.hash !== undefined) {
            writer.uint32(26).string(message.hash);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseExecutionContext();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.threadId = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.processId = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.hash = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            threadId: isSet(object.threadId) ? globalThis.String(object.threadId) : "",
            processId: isSet(object.processId) ? globalThis.String(object.processId) : undefined,
            hash: isSet(object.hash) ? globalThis.String(object.hash) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.threadId !== "") {
            obj.threadId = message.threadId;
        }
        if (message.processId !== undefined) {
            obj.processId = message.processId;
        }
        if (message.hash !== undefined) {
            obj.hash = message.hash;
        }
        return obj;
    },
};
function createBaseTestFrameworkEventResponse() {
    return { success: false, error: undefined };
}
export const TestFrameworkEventResponse = {
    encode(message, writer = new BinaryWriter()) {
        if (message.success !== false) {
            writer.uint32(8).bool(message.success);
        }
        if (message.error !== undefined) {
            writer.uint32(18).string(message.error);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTestFrameworkEventResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.success = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.error = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            success: isSet(object.success) ? globalThis.Boolean(object.success) : false,
            error: isSet(object.error) ? globalThis.String(object.error) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.success !== false) {
            obj.success = message.success;
        }
        if (message.error !== undefined) {
            obj.error = message.error;
        }
        return obj;
    },
};
function createBaseEnqueueTestEventRequest() {
    return { binSessionId: "", eventUrl: "", eventDataJson: new Uint8Array(0) };
}
export const EnqueueTestEventRequest = {
    encode(message, writer = new BinaryWriter()) {
        if (message.binSessionId !== "") {
            writer.uint32(10).string(message.binSessionId);
        }
        if (message.eventUrl !== "") {
            writer.uint32(18).string(message.eventUrl);
        }
        if (message.eventDataJson.length !== 0) {
            writer.uint32(26).bytes(message.eventDataJson);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseEnqueueTestEventRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.binSessionId = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.eventUrl = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.eventDataJson = reader.bytes();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            binSessionId: isSet(object.binSessionId) ? globalThis.String(object.binSessionId) : "",
            eventUrl: isSet(object.eventUrl) ? globalThis.String(object.eventUrl) : "",
            eventDataJson: isSet(object.eventDataJson) ? bytesFromBase64(object.eventDataJson) : new Uint8Array(0),
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.binSessionId !== "") {
            obj.binSessionId = message.binSessionId;
        }
        if (message.eventUrl !== "") {
            obj.eventUrl = message.eventUrl;
        }
        if (message.eventDataJson.length !== 0) {
            obj.eventDataJson = base64FromBytes(message.eventDataJson);
        }
        return obj;
    },
};
function createBaseEnqueueTestEventResponse() {
    return { success: false, error: undefined };
}
export const EnqueueTestEventResponse = {
    encode(message, writer = new BinaryWriter()) {
        if (message.success !== false) {
            writer.uint32(8).bool(message.success);
        }
        if (message.error !== undefined) {
            writer.uint32(18).string(message.error);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseEnqueueTestEventResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.success = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.error = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            success: isSet(object.success) ? globalThis.Boolean(object.success) : false,
            error: isSet(object.error) ? globalThis.String(object.error) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.success !== false) {
            obj.success = message.success;
        }
        if (message.error !== undefined) {
            obj.error = message.error;
        }
        return obj;
    },
};
function createBaseLogCreatedEventRequest() {
    return { binSessionId: "", platformIndex: 0, logs: [], executionContext: undefined };
}
export const LogCreatedEventRequest = {
    encode(message, writer = new BinaryWriter()) {
        if (message.binSessionId !== "") {
            writer.uint32(10).string(message.binSessionId);
        }
        if (message.platformIndex !== 0) {
            writer.uint32(16).uint32(message.platformIndex);
        }
        for (const v of message.logs) {
            LogCreatedEventRequest_LogEntry.encode(v, writer.uint32(26).fork()).join();
        }
        if (message.executionContext !== undefined) {
            ExecutionContext.encode(message.executionContext, writer.uint32(34).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseLogCreatedEventRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.binSessionId = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.platformIndex = reader.uint32();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.logs.push(LogCreatedEventRequest_LogEntry.decode(reader, reader.uint32()));
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.executionContext = ExecutionContext.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            binSessionId: isSet(object.binSessionId) ? globalThis.String(object.binSessionId) : "",
            platformIndex: isSet(object.platformIndex) ? globalThis.Number(object.platformIndex) : 0,
            logs: globalThis.Array.isArray(object?.logs)
                ? object.logs.map((e) => LogCreatedEventRequest_LogEntry.fromJSON(e))
                : [],
            executionContext: isSet(object.executionContext) ? ExecutionContext.fromJSON(object.executionContext) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.binSessionId !== "") {
            obj.binSessionId = message.binSessionId;
        }
        if (message.platformIndex !== 0) {
            obj.platformIndex = Math.round(message.platformIndex);
        }
        if (message.logs?.length) {
            obj.logs = message.logs.map((e) => LogCreatedEventRequest_LogEntry.toJSON(e));
        }
        if (message.executionContext !== undefined) {
            obj.executionContext = ExecutionContext.toJSON(message.executionContext);
        }
        return obj;
    },
};
function createBaseLogCreatedEventRequest_LogEntry() {
    return {
        testFrameworkName: "",
        testFrameworkVersion: "",
        testFrameworkState: "",
        timestamp: "",
        uuid: "",
        kind: "",
        message: new Uint8Array(0),
        level: undefined,
        fileName: undefined,
        fileSize: undefined,
        filePath: undefined,
    };
}
export const LogCreatedEventRequest_LogEntry = {
    encode(message, writer = new BinaryWriter()) {
        if (message.testFrameworkName !== "") {
            writer.uint32(10).string(message.testFrameworkName);
        }
        if (message.testFrameworkVersion !== "") {
            writer.uint32(18).string(message.testFrameworkVersion);
        }
        if (message.testFrameworkState !== "") {
            writer.uint32(26).string(message.testFrameworkState);
        }
        if (message.timestamp !== "") {
            writer.uint32(34).string(message.timestamp);
        }
        if (message.uuid !== "") {
            writer.uint32(42).string(message.uuid);
        }
        if (message.kind !== "") {
            writer.uint32(50).string(message.kind);
        }
        if (message.message.length !== 0) {
            writer.uint32(58).bytes(message.message);
        }
        if (message.level !== undefined) {
            writer.uint32(66).string(message.level);
        }
        if (message.fileName !== undefined) {
            writer.uint32(74).string(message.fileName);
        }
        if (message.fileSize !== undefined) {
            writer.uint32(80).int64(message.fileSize);
        }
        if (message.filePath !== undefined) {
            writer.uint32(90).string(message.filePath);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseLogCreatedEventRequest_LogEntry();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.testFrameworkName = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.testFrameworkVersion = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.testFrameworkState = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.timestamp = reader.string();
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.uuid = reader.string();
                    continue;
                }
                case 6: {
                    if (tag !== 50) {
                        break;
                    }
                    message.kind = reader.string();
                    continue;
                }
                case 7: {
                    if (tag !== 58) {
                        break;
                    }
                    message.message = reader.bytes();
                    continue;
                }
                case 8: {
                    if (tag !== 66) {
                        break;
                    }
                    message.level = reader.string();
                    continue;
                }
                case 9: {
                    if (tag !== 74) {
                        break;
                    }
                    message.fileName = reader.string();
                    continue;
                }
                case 10: {
                    if (tag !== 80) {
                        break;
                    }
                    message.fileSize = longToNumber(reader.int64());
                    continue;
                }
                case 11: {
                    if (tag !== 90) {
                        break;
                    }
                    message.filePath = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            testFrameworkName: isSet(object.testFrameworkName) ? globalThis.String(object.testFrameworkName) : "",
            testFrameworkVersion: isSet(object.testFrameworkVersion) ? globalThis.String(object.testFrameworkVersion) : "",
            testFrameworkState: isSet(object.testFrameworkState) ? globalThis.String(object.testFrameworkState) : "",
            timestamp: isSet(object.timestamp) ? globalThis.String(object.timestamp) : "",
            uuid: isSet(object.uuid) ? globalThis.String(object.uuid) : "",
            kind: isSet(object.kind) ? globalThis.String(object.kind) : "",
            message: isSet(object.message) ? bytesFromBase64(object.message) : new Uint8Array(0),
            level: isSet(object.level) ? globalThis.String(object.level) : undefined,
            fileName: isSet(object.fileName) ? globalThis.String(object.fileName) : undefined,
            fileSize: isSet(object.fileSize) ? globalThis.Number(object.fileSize) : undefined,
            filePath: isSet(object.filePath) ? globalThis.String(object.filePath) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.testFrameworkName !== "") {
            obj.testFrameworkName = message.testFrameworkName;
        }
        if (message.testFrameworkVersion !== "") {
            obj.testFrameworkVersion = message.testFrameworkVersion;
        }
        if (message.testFrameworkState !== "") {
            obj.testFrameworkState = message.testFrameworkState;
        }
        if (message.timestamp !== "") {
            obj.timestamp = message.timestamp;
        }
        if (message.uuid !== "") {
            obj.uuid = message.uuid;
        }
        if (message.kind !== "") {
            obj.kind = message.kind;
        }
        if (message.message.length !== 0) {
            obj.message = base64FromBytes(message.message);
        }
        if (message.level !== undefined) {
            obj.level = message.level;
        }
        if (message.fileName !== undefined) {
            obj.fileName = message.fileName;
        }
        if (message.fileSize !== undefined) {
            obj.fileSize = Math.round(message.fileSize);
        }
        if (message.filePath !== undefined) {
            obj.filePath = message.filePath;
        }
        return obj;
    },
};
function createBaseLogCreatedEventResponse() {
    return { success: false, error: undefined };
}
export const LogCreatedEventResponse = {
    encode(message, writer = new BinaryWriter()) {
        if (message.success !== false) {
            writer.uint32(8).bool(message.success);
        }
        if (message.error !== undefined) {
            writer.uint32(18).string(message.error);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseLogCreatedEventResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.success = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.error = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            success: isSet(object.success) ? globalThis.Boolean(object.success) : false,
            error: isSet(object.error) ? globalThis.String(object.error) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.success !== false) {
            obj.success = message.success;
        }
        if (message.error !== undefined) {
            obj.error = message.error;
        }
        return obj;
    },
};
function createBaseTestSessionEventRequest() {
    return {
        binSessionId: "",
        platformIndex: 0,
        testFrameworkName: "",
        testFrameworkVersion: "",
        testFrameworkState: "",
        testHookState: "",
        testUuid: "",
        automationSessions: [],
        executionContext: undefined,
        capabilities: undefined,
    };
}
export const TestSessionEventRequest = {
    encode(message, writer = new BinaryWriter()) {
        if (message.binSessionId !== "") {
            writer.uint32(10).string(message.binSessionId);
        }
        if (message.platformIndex !== 0) {
            writer.uint32(16).uint32(message.platformIndex);
        }
        if (message.testFrameworkName !== "") {
            writer.uint32(26).string(message.testFrameworkName);
        }
        if (message.testFrameworkVersion !== "") {
            writer.uint32(34).string(message.testFrameworkVersion);
        }
        if (message.testFrameworkState !== "") {
            writer.uint32(42).string(message.testFrameworkState);
        }
        if (message.testHookState !== "") {
            writer.uint32(50).string(message.testHookState);
        }
        if (message.testUuid !== "") {
            writer.uint32(58).string(message.testUuid);
        }
        for (const v of message.automationSessions) {
            TestSessionEventRequest_AutomationSession.encode(v, writer.uint32(66).fork()).join();
        }
        if (message.executionContext !== undefined) {
            ExecutionContext.encode(message.executionContext, writer.uint32(74).fork()).join();
        }
        if (message.capabilities !== undefined) {
            writer.uint32(82).bytes(message.capabilities);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTestSessionEventRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.binSessionId = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 16) {
                        break;
                    }
                    message.platformIndex = reader.uint32();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.testFrameworkName = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.testFrameworkVersion = reader.string();
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.testFrameworkState = reader.string();
                    continue;
                }
                case 6: {
                    if (tag !== 50) {
                        break;
                    }
                    message.testHookState = reader.string();
                    continue;
                }
                case 7: {
                    if (tag !== 58) {
                        break;
                    }
                    message.testUuid = reader.string();
                    continue;
                }
                case 8: {
                    if (tag !== 66) {
                        break;
                    }
                    message.automationSessions.push(TestSessionEventRequest_AutomationSession.decode(reader, reader.uint32()));
                    continue;
                }
                case 9: {
                    if (tag !== 74) {
                        break;
                    }
                    message.executionContext = ExecutionContext.decode(reader, reader.uint32());
                    continue;
                }
                case 10: {
                    if (tag !== 82) {
                        break;
                    }
                    message.capabilities = reader.bytes();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            binSessionId: isSet(object.binSessionId) ? globalThis.String(object.binSessionId) : "",
            platformIndex: isSet(object.platformIndex) ? globalThis.Number(object.platformIndex) : 0,
            testFrameworkName: isSet(object.testFrameworkName) ? globalThis.String(object.testFrameworkName) : "",
            testFrameworkVersion: isSet(object.testFrameworkVersion) ? globalThis.String(object.testFrameworkVersion) : "",
            testFrameworkState: isSet(object.testFrameworkState) ? globalThis.String(object.testFrameworkState) : "",
            testHookState: isSet(object.testHookState) ? globalThis.String(object.testHookState) : "",
            testUuid: isSet(object.testUuid) ? globalThis.String(object.testUuid) : "",
            automationSessions: globalThis.Array.isArray(object?.automationSessions)
                ? object.automationSessions.map((e) => TestSessionEventRequest_AutomationSession.fromJSON(e))
                : [],
            executionContext: isSet(object.executionContext) ? ExecutionContext.fromJSON(object.executionContext) : undefined,
            capabilities: isSet(object.capabilities) ? bytesFromBase64(object.capabilities) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.binSessionId !== "") {
            obj.binSessionId = message.binSessionId;
        }
        if (message.platformIndex !== 0) {
            obj.platformIndex = Math.round(message.platformIndex);
        }
        if (message.testFrameworkName !== "") {
            obj.testFrameworkName = message.testFrameworkName;
        }
        if (message.testFrameworkVersion !== "") {
            obj.testFrameworkVersion = message.testFrameworkVersion;
        }
        if (message.testFrameworkState !== "") {
            obj.testFrameworkState = message.testFrameworkState;
        }
        if (message.testHookState !== "") {
            obj.testHookState = message.testHookState;
        }
        if (message.testUuid !== "") {
            obj.testUuid = message.testUuid;
        }
        if (message.automationSessions?.length) {
            obj.automationSessions = message.automationSessions.map((e) => TestSessionEventRequest_AutomationSession.toJSON(e));
        }
        if (message.executionContext !== undefined) {
            obj.executionContext = ExecutionContext.toJSON(message.executionContext);
        }
        if (message.capabilities !== undefined) {
            obj.capabilities = base64FromBytes(message.capabilities);
        }
        return obj;
    },
};
function createBaseTestSessionEventRequest_AutomationSession() {
    return {
        provider: "",
        product: undefined,
        frameworkName: "",
        frameworkVersion: "",
        frameworkSessionId: "",
        ref: undefined,
        hubUrl: undefined,
    };
}
export const TestSessionEventRequest_AutomationSession = {
    encode(message, writer = new BinaryWriter()) {
        if (message.provider !== "") {
            writer.uint32(10).string(message.provider);
        }
        if (message.product !== undefined) {
            writer.uint32(18).string(message.product);
        }
        if (message.frameworkName !== "") {
            writer.uint32(26).string(message.frameworkName);
        }
        if (message.frameworkVersion !== "") {
            writer.uint32(34).string(message.frameworkVersion);
        }
        if (message.frameworkSessionId !== "") {
            writer.uint32(42).string(message.frameworkSessionId);
        }
        if (message.ref !== undefined) {
            writer.uint32(50).string(message.ref);
        }
        if (message.hubUrl !== undefined) {
            writer.uint32(58).string(message.hubUrl);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTestSessionEventRequest_AutomationSession();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.provider = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.product = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.frameworkName = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.frameworkVersion = reader.string();
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.frameworkSessionId = reader.string();
                    continue;
                }
                case 6: {
                    if (tag !== 50) {
                        break;
                    }
                    message.ref = reader.string();
                    continue;
                }
                case 7: {
                    if (tag !== 58) {
                        break;
                    }
                    message.hubUrl = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            provider: isSet(object.provider) ? globalThis.String(object.provider) : "",
            product: isSet(object.product) ? globalThis.String(object.product) : undefined,
            frameworkName: isSet(object.frameworkName) ? globalThis.String(object.frameworkName) : "",
            frameworkVersion: isSet(object.frameworkVersion) ? globalThis.String(object.frameworkVersion) : "",
            frameworkSessionId: isSet(object.frameworkSessionId) ? globalThis.String(object.frameworkSessionId) : "",
            ref: isSet(object.ref) ? globalThis.String(object.ref) : undefined,
            hubUrl: isSet(object.hubUrl) ? globalThis.String(object.hubUrl) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.provider !== "") {
            obj.provider = message.provider;
        }
        if (message.product !== undefined) {
            obj.product = message.product;
        }
        if (message.frameworkName !== "") {
            obj.frameworkName = message.frameworkName;
        }
        if (message.frameworkVersion !== "") {
            obj.frameworkVersion = message.frameworkVersion;
        }
        if (message.frameworkSessionId !== "") {
            obj.frameworkSessionId = message.frameworkSessionId;
        }
        if (message.ref !== undefined) {
            obj.ref = message.ref;
        }
        if (message.hubUrl !== undefined) {
            obj.hubUrl = message.hubUrl;
        }
        return obj;
    },
};
function createBaseTestSessionEventResponse() {
    return { success: false, error: undefined };
}
export const TestSessionEventResponse = {
    encode(message, writer = new BinaryWriter()) {
        if (message.success !== false) {
            writer.uint32(8).bool(message.success);
        }
        if (message.error !== undefined) {
            writer.uint32(18).string(message.error);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTestSessionEventResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.success = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.error = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            success: isSet(object.success) ? globalThis.Boolean(object.success) : false,
            error: isSet(object.error) ? globalThis.String(object.error) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.success !== false) {
            obj.success = message.success;
        }
        if (message.error !== undefined) {
            obj.error = message.error;
        }
        return obj;
    },
};
function createBaseFetchDriverExecuteParamsEventRequest() {
    return { binSessionId: "", product: "", scriptName: undefined };
}
export const FetchDriverExecuteParamsEventRequest = {
    encode(message, writer = new BinaryWriter()) {
        if (message.binSessionId !== "") {
            writer.uint32(10).string(message.binSessionId);
        }
        if (message.product !== "") {
            writer.uint32(18).string(message.product);
        }
        if (message.scriptName !== undefined) {
            writer.uint32(26).string(message.scriptName);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseFetchDriverExecuteParamsEventRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.binSessionId = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.product = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.scriptName = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            binSessionId: isSet(object.binSessionId) ? globalThis.String(object.binSessionId) : "",
            product: isSet(object.product) ? globalThis.String(object.product) : "",
            scriptName: isSet(object.scriptName) ? globalThis.String(object.scriptName) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.binSessionId !== "") {
            obj.binSessionId = message.binSessionId;
        }
        if (message.product !== "") {
            obj.product = message.product;
        }
        if (message.scriptName !== undefined) {
            obj.scriptName = message.scriptName;
        }
        return obj;
    },
};
function createBaseFetchDriverExecuteParamsEventResponse() {
    return { success: false, script: undefined, accessibilityExecuteParams: undefined, error: undefined };
}
export const FetchDriverExecuteParamsEventResponse = {
    encode(message, writer = new BinaryWriter()) {
        if (message.success !== false) {
            writer.uint32(8).bool(message.success);
        }
        if (message.script !== undefined) {
            writer.uint32(18).string(message.script);
        }
        if (message.accessibilityExecuteParams !== undefined) {
            writer.uint32(26).bytes(message.accessibilityExecuteParams);
        }
        if (message.error !== undefined) {
            writer.uint32(34).string(message.error);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseFetchDriverExecuteParamsEventResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.success = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.script = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.accessibilityExecuteParams = reader.bytes();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.error = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            success: isSet(object.success) ? globalThis.Boolean(object.success) : false,
            script: isSet(object.script) ? globalThis.String(object.script) : undefined,
            accessibilityExecuteParams: isSet(object.accessibilityExecuteParams)
                ? bytesFromBase64(object.accessibilityExecuteParams)
                : undefined,
            error: isSet(object.error) ? globalThis.String(object.error) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.success !== false) {
            obj.success = message.success;
        }
        if (message.script !== undefined) {
            obj.script = message.script;
        }
        if (message.accessibilityExecuteParams !== undefined) {
            obj.accessibilityExecuteParams = base64FromBytes(message.accessibilityExecuteParams);
        }
        if (message.error !== undefined) {
            obj.error = message.error;
        }
        return obj;
    },
};
function bytesFromBase64(b64) {
    if (globalThis.Buffer) {
        return Uint8Array.from(globalThis.Buffer.from(b64, "base64"));
    }
    else {
        const bin = globalThis.atob(b64);
        const arr = new Uint8Array(bin.length);
        for (let i = 0; i < bin.length; ++i) {
            arr[i] = bin.charCodeAt(i);
        }
        return arr;
    }
}
function base64FromBytes(arr) {
    if (globalThis.Buffer) {
        return globalThis.Buffer.from(arr).toString("base64");
    }
    else {
        const bin = [];
        arr.forEach((byte) => {
            bin.push(globalThis.String.fromCharCode(byte));
        });
        return globalThis.btoa(bin.join(""));
    }
}
function longToNumber(int64) {
    const num = globalThis.Number(int64.toString());
    if (num > globalThis.Number.MAX_SAFE_INTEGER) {
        throw new globalThis.Error("Value is larger than Number.MAX_SAFE_INTEGER");
    }
    if (num < globalThis.Number.MIN_SAFE_INTEGER) {
        throw new globalThis.Error("Value is smaller than Number.MIN_SAFE_INTEGER");
    }
    return num;
}
function isObject(value) {
    return typeof value === "object" && value !== null;
}
function isSet(value) {
    return value !== null && value !== undefined;
}
