// @browserstack/wdio-browserstack-service/dist/generated/sdk-messages-ai.js
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.7.7
//   protoc               unknown
// source: sdk-messages-ai.proto
/* eslint-disable */
import { BinaryReader, BinaryWriter } from "@bufbuild/protobuf/wire";
export const protobufPackage = "browserstack.sdk.v1";
function createBaseAIBrowserExtensionRequest() {
    return { binSessionId: "", browserName: "", browserVersion: "" };
}
export const AIBrowserExtensionRequest = {
    encode(message, writer = new BinaryWriter()) {
        if (message.binSessionId !== "") {
            writer.uint32(10).string(message.binSessionId);
        }
        if (message.browserName !== "") {
            writer.uint32(18).string(message.browserName);
        }
        if (message.browserVersion !== "") {
            writer.uint32(26).string(message.browserVersion);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAIBrowserExtensionRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.binSessionId = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.browserName = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.browserVersion = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            binSessionId: isSet(object.binSessionId) ? globalThis.String(object.binSessionId) : "",
            browserName: isSet(object.browserName) ? globalThis.String(object.browserName) : "",
            browserVersion: isSet(object.browserVersion) ? globalThis.String(object.browserVersion) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.binSessionId !== "") {
            obj.binSessionId = message.binSessionId;
        }
        if (message.browserName !== "") {
            obj.browserName = message.browserName;
        }
        if (message.browserVersion !== "") {
            obj.browserVersion = message.browserVersion;
        }
        return obj;
    },
};
function createBaseAIBrowserExtensionResponse() {
    return { success: false, extension: undefined, error: undefined };
}
export const AIBrowserExtensionResponse = {
    encode(message, writer = new BinaryWriter()) {
        if (message.success !== false) {
            writer.uint32(8).bool(message.success);
        }
        if (message.extension !== undefined) {
            writer.uint32(18).string(message.extension);
        }
        if (message.error !== undefined) {
            writer.uint32(26).string(message.error);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAIBrowserExtensionResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.success = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.extension = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.error = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            success: isSet(object.success) ? globalThis.Boolean(object.success) : false,
            extension: isSet(object.extension) ? globalThis.String(object.extension) : undefined,
            error: isSet(object.error) ? globalThis.String(object.error) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.success !== false) {
            obj.success = message.success;
        }
        if (message.extension !== undefined) {
            obj.extension = message.extension;
        }
        if (message.error !== undefined) {
            obj.error = message.error;
        }
        return obj;
    },
};
function createBaseAISelfHealGetRequest() {
    return { binSessionId: "", platformIndex: undefined, frameworkSessionId: "", locatorType: "" };
}
export const AISelfHealGetRequest = {
    encode(message, writer = new BinaryWriter()) {
        if (message.binSessionId !== "") {
            writer.uint32(10).string(message.binSessionId);
        }
        if (message.platformIndex !== undefined) {
            writer.uint32(18).string(message.platformIndex);
        }
        if (message.frameworkSessionId !== "") {
            writer.uint32(26).string(message.frameworkSessionId);
        }
        if (message.locatorType !== "") {
            writer.uint32(34).string(message.locatorType);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAISelfHealGetRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.binSessionId = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.platformIndex = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.frameworkSessionId = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.locatorType = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            binSessionId: isSet(object.binSessionId) ? globalThis.String(object.binSessionId) : "",
            platformIndex: isSet(object.platformIndex) ? globalThis.String(object.platformIndex) : undefined,
            frameworkSessionId: isSet(object.frameworkSessionId) ? globalThis.String(object.frameworkSessionId) : "",
            locatorType: isSet(object.locatorType) ? globalThis.String(object.locatorType) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.binSessionId !== "") {
            obj.binSessionId = message.binSessionId;
        }
        if (message.platformIndex !== undefined) {
            obj.platformIndex = message.platformIndex;
        }
        if (message.frameworkSessionId !== "") {
            obj.frameworkSessionId = message.frameworkSessionId;
        }
        if (message.locatorType !== "") {
            obj.locatorType = message.locatorType;
        }
        return obj;
    },
};
function createBaseAISelfHealGetResponse() {
    return {
        success: false,
        frameworkSessionId: "",
        locatorType: undefined,
        locatorValue: undefined,
        aiHealing: undefined,
        error: undefined,
    };
}
export const AISelfHealGetResponse = {
    encode(message, writer = new BinaryWriter()) {
        if (message.success !== false) {
            writer.uint32(8).bool(message.success);
        }
        if (message.frameworkSessionId !== "") {
            writer.uint32(18).string(message.frameworkSessionId);
        }
        if (message.locatorType !== undefined) {
            writer.uint32(26).string(message.locatorType);
        }
        if (message.locatorValue !== undefined) {
            writer.uint32(34).string(message.locatorValue);
        }
        if (message.aiHealing !== undefined) {
            writer.uint32(40).bool(message.aiHealing);
        }
        if (message.error !== undefined) {
            writer.uint32(50).string(message.error);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAISelfHealGetResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 8) {
                        break;
                    }
                    message.success = reader.bool();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.frameworkSessionId = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.locatorType = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.locatorValue = reader.string();
                    continue;
                }
                case 5: {
                    if (tag !== 40) {
                        break;
                    }
                    message.aiHealing = reader.bool();
                    continue;
                }
                case 6: {
                    if (tag !== 50) {
                        break;
                    }
                    message.error = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            success: isSet(object.success) ? globalThis.Boolean(object.success) : false,
            frameworkSessionId: isSet(object.frameworkSessionId) ? globalThis.String(object.frameworkSessionId) : "",
            locatorType: isSet(object.locatorType) ? globalThis.String(object.locatorType) : undefined,
            locatorValue: isSet(object.locatorValue) ? globalThis.String(object.locatorValue) : undefined,
            aiHealing: isSet(object.aiHealing) ? globalThis.Boolean(object.aiHealing) : undefined,
            error: isSet(object.error) ? globalThis.String(object.error) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.success !== false) {
            obj.success = message.success;
        }
        if (message.frameworkSessionId !== "") {
            obj.frameworkSessionId = message.frameworkSessionId;
        }
        if (message.locatorType !== undefined) {
            obj.locatorType = message.locatorType;
        }
        if (message.locatorValue !== undefined) {
            obj.locatorValue = message.locatorValue;
        }
        if (message.aiHealing !== undefined) {
            obj.aiHealing = message.aiHealing;
        }
        if (message.error !== undefined) {
            obj.error = message.error;
        }
        return obj;
    },
};
function createBaseAISelfHealStepRequest() {
    return {
        binSessionId: "",
        platformIndex: undefined,
        frameworkSessionId: "",
        isSuccess: false,
        locatorType: "",
        locatorValue: "",
        testName: "",
    };
}
export const AISelfHealStepRequest = {
    encode(message, writer = new BinaryWriter()) {
        if (message.binSessionId !== "") {
            writer.uint32(10).string(message.binSessionId);
        }
        if (message.platformIndex !== undefined) {
            writer.uint32(18).string(message.platformIndex);
        }
        if (message.frameworkSessionId !== "") {
            writer.uint32(26).string(message.frameworkSessionId);
        }
        if (message.isSuccess !== false) {
            writer.uint32(32).bool(message.isSuccess);
        }
        if (message.locatorType !== "") {
            writer.uint32(42).string(message.locatorType);
        }
        if (message.locatorValue !== "") {
            writer.uint32(50).string(message.locatorValue);
        }
        if (message.testName !== "") {
            writer.uint32(58).string(message.testName);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAISelfHealStepRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.binSessionId = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.platformIndex = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.frameworkSessionId = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 32) {
                        break;
                    }
                    message.isSuccess = reader.bool();
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.locatorType = reader.string();
                    continue;
                }
                case 6: {
                    if (tag !== 50) {
                        break;
                    }
                    message.locatorValue = reader.string();
                    continue;
                }
                case 7: {
                    if (tag !== 58) {
                        break;
                    }
                    message.testName = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            binSessionId: isSet(object.binSessionId) ? globalThis.String(object.binSessionId) : "",
            platformIndex: isSet(object.platformIndex) ? globalThis.String(object.platformIndex) : undefined,
            frameworkSessionId: isSet(object.frameworkSessionId) ? globalThis.String(object.frameworkSessionId) : "",
            isSuccess: isSet(object.isSuccess) ? globalThis.Boolean(object.isSuccess) : false,
            locatorType: isSet(object.locatorType) ? globalThis.String(object.locatorType) : "",
            locatorValue: isSet(object.locatorValue) ? globalThis.String(object.locatorValue) : "",
            testName: isSet(object.testName) ? globalThis.String(object.testName) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.binSessionId !== "") {
            obj.binSessionId = message.binSessionId;
        }
        if (message.platformIndex !== undefined) {
            obj.platformIndex = message.platformIndex;
        }
        if (message.frameworkSessionId !== "") {
            obj.frameworkSessionId = message.frameworkSessionId;
        }
        if (message.isSuccess !== false) {
            obj.isSuccess = message.isSuccess;
        }
        if (message.locatorType !== "") {
            obj.locatorType = message.locatorType;
        }
        if (message.locatorValue !== "") {
            obj.locatorValue = message.locatorValue;
        }
        if (message.testName !== "") {
            obj.testName = message.testName;
        }
        return obj;
    },
};
function createBaseAISelfHealStepResponse() {
    return { frameworkSessionId: "", executeScript: "" };
}
export const AISelfHealStepResponse = {
    encode(message, writer = new BinaryWriter()) {
        if (message.frameworkSessionId !== "") {
            writer.uint32(10).string(message.frameworkSessionId);
        }
        if (message.executeScript !== "") {
            writer.uint32(18).string(message.executeScript);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseAISelfHealStepResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.frameworkSessionId = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.executeScript = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            frameworkSessionId: isSet(object.frameworkSessionId) ? globalThis.String(object.frameworkSessionId) : "",
            executeScript: isSet(object.executeScript) ? globalThis.String(object.executeScript) : "",
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.frameworkSessionId !== "") {
            obj.frameworkSessionId = message.frameworkSessionId;
        }
        if (message.executeScript !== "") {
            obj.executeScript = message.executeScript;
        }
        return obj;
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
