// @browserstack/wdio-browserstack-service/dist/generated/sdk-messages-testhub.js
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.7.7
//   protoc               unknown
// source: sdk-messages-testhub.proto
/* eslint-disable */
import { BinaryReader, BinaryWriter } from "@bufbuild/protobuf/wire";
export const protobufPackage = "browserstack.sdk.v1";
function createBaseTestHub() {
    return { jwt: "", buildHashedId: "", testhubEvents: false, errors: new Uint8Array(0) };
}
export const TestHub = {
    encode(message, writer = new BinaryWriter()) {
        if (message.jwt !== "") {
            writer.uint32(10).string(message.jwt);
        }
        if (message.buildHashedId !== "") {
            writer.uint32(18).string(message.buildHashedId);
        }
        if (message.testhubEvents !== false) {
            writer.uint32(24).bool(message.testhubEvents);
        }
        if (message.errors.length !== 0) {
            writer.uint32(34).bytes(message.errors);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseTestHub();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.jwt = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.buildHashedId = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 24) {
                        break;
                    }
                    message.testhubEvents = reader.bool();
                    continue;
                }
                case 4: {
                    if (tag !== 34) {
                        break;
                    }
                    message.errors = reader.bytes();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            jwt: isSet(object.jwt) ? globalThis.String(object.jwt) : "",
            buildHashedId: isSet(object.buildHashedId) ? globalThis.String(object.buildHashedId) : "",
            testhubEvents: isSet(object.testhubEvents) ? globalThis.Boolean(object.testhubEvents) : false,
            errors: isSet(object.errors) ? bytesFromBase64(object.errors) : new Uint8Array(0),
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.jwt !== "") {
            obj.jwt = message.jwt;
        }
        if (message.buildHashedId !== "") {
            obj.buildHashedId = message.buildHashedId;
        }
        if (message.testhubEvents !== false) {
            obj.testhubEvents = message.testhubEvents;
        }
        if (message.errors.length !== 0) {
            obj.errors = base64FromBytes(message.errors);
        }
        return obj;
    },
};
function bytesFromBase64(b64) {
    if (globalThis.Buffer) {
        return Uint8Array.from(globalThis.Buffer.from(b64, "base64"));
    }
    else {
        const bin = globalThis.atob(b64);
        const arr = new Uint8Array(bin.length);
        for (let i = 0; i < bin.length; ++i) {
            arr[i] = bin.charCodeAt(i);
        }
        return arr;
    }
}
function base64FromBytes(arr) {
    if (globalThis.Buffer) {
        return globalThis.Buffer.from(arr).toString("base64");
    }
    else {
        const bin = [];
        arr.forEach((byte) => {
            bin.push(globalThis.String.fromCharCode(byte));
        });
        return globalThis.btoa(bin.join(""));
    }
}
function isSet(value) {
    return value !== null && value !== undefined;
}
