// @browserstack/wdio-browserstack-service/dist/generated/sdk.js
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.7.7
//   protoc               unknown
// source: sdk.proto
/* eslint-disable */
import { makeGenericClientConstructor, } from "@grpc/grpc-js";
import { AccessibilityConfigRequest, AccessibilityConfigResponse, AccessibilityResultRequest, AccessibilityResultResponse, } from "./sdk-messages-accessibility.js";
import { AIBrowserExtensionRequest, AIBrowserExtensionResponse, AISelfHealGetRequest, AISelfHealGetResponse, AISelfHealStepRequest, AISelfHealStepResponse, } from "./sdk-messages-ai.js";
import { ObservabilityConfigRequest, ObservabilityConfigResponse } from "./sdk-messages-observability.js";
import { AutomationFrameworkInitRequest, AutomationFrameworkInitResponse, AutomationFrameworkStartRequest, AutomationFrameworkStartResponse, AutomationFrameworkStopRequest, AutomationFrameworkStopResponse, ConnectBinSessionRequest, ConnectBinSessionResponse, DriverInitRequest, DriverInitResponse, EnqueueTestEventRequest, EnqueueTestEventResponse, FetchDriverExecuteParamsEventRequest, FetchDriverExecuteParamsEventResponse, FindNearestHubRequest, FindNearestHubResponse, LogCreatedEventRequest, LogCreatedEventResponse, StartBinSessionRequest, StartBinSessionResponse, StopBinSessionRequest, StopBinSessionResponse, TestFrameworkEventRequest, TestFrameworkEventResponse, TestOrchestrationRequest, TestOrchestrationResponse, TestSessionEventRequest, TestSessionEventResponse, } from "./sdk-messages.js";
export const protobufPackage = "browserstack.sdk.v1";
export const SDKService = {
    startBinSession: {
        path: "/browserstack.sdk.v1.SDK/StartBinSession",
        requestStream: false,
        responseStream: false,
        requestSerialize: (value) => Buffer.from(StartBinSessionRequest.encode(value).finish()),
        requestDeserialize: (value) => StartBinSessionRequest.decode(value),
        responseSerialize: (value) => Buffer.from(StartBinSessionResponse.encode(value).finish()),
        responseDeserialize: (value) => StartBinSessionResponse.decode(value),
    },
    connectBinSession: {
        path: "/browserstack.sdk.v1.SDK/ConnectBinSession",
        requestStream: false,
        responseStream: false,
        requestSerialize: (value) => Buffer.from(ConnectBinSessionRequest.encode(value).finish()),
        requestDeserialize: (value) => ConnectBinSessionRequest.decode(value),
        responseSerialize: (value) => Buffer.from(ConnectBinSessionResponse.encode(value).finish()),
        responseDeserialize: (value) => ConnectBinSessionResponse.decode(value),
    },
    stopBinSession: {
        path: "/browserstack.sdk.v1.SDK/StopBinSession",
        requestStream: false,
        responseStream: false,
        requestSerialize: (value) => Buffer.from(StopBinSessionRequest.encode(value).finish()),
        requestDeserialize: (value) => StopBinSessionRequest.decode(value),
        responseSerialize: (value) => Buffer.from(StopBinSessionResponse.encode(value).finish()),
        responseDeserialize: (value) => StopBinSessionResponse.decode(value),
    },
    driverInit: {
        path: "/browserstack.sdk.v1.SDK/DriverInit",
        requestStream: false,
        responseStream: false,
        requestSerialize: (value) => Buffer.from(DriverInitRequest.encode(value).finish()),
        requestDeserialize: (value) => DriverInitRequest.decode(value),
        responseSerialize: (value) => Buffer.from(DriverInitResponse.encode(value).finish()),
        responseDeserialize: (value) => DriverInitResponse.decode(value),
    },
    automationFrameworkInit: {
        path: "/browserstack.sdk.v1.SDK/AutomationFrameworkInit",
        requestStream: false,
        responseStream: false,
        requestSerialize: (value) => Buffer.from(AutomationFrameworkInitRequest.encode(value).finish()),
        requestDeserialize: (value) => AutomationFrameworkInitRequest.decode(value),
        responseSerialize: (value) => Buffer.from(AutomationFrameworkInitResponse.encode(value).finish()),
        responseDeserialize: (value) => AutomationFrameworkInitResponse.decode(value),
    },
    /** session-start */
    automationFrameworkStart: {
        path: "/browserstack.sdk.v1.SDK/AutomationFrameworkStart",
        requestStream: false,
        responseStream: false,
        requestSerialize: (value) => Buffer.from(AutomationFrameworkStartRequest.encode(value).finish()),
        requestDeserialize: (value) => AutomationFrameworkStartRequest.decode(value),
        responseSerialize: (value) => Buffer.from(AutomationFrameworkStartResponse.encode(value).finish()),
        responseDeserialize: (value) => AutomationFrameworkStartResponse.decode(value),
    },
    automationFrameworkStop: {
        path: "/browserstack.sdk.v1.SDK/AutomationFrameworkStop",
        requestStream: false,
        responseStream: false,
        requestSerialize: (value) => Buffer.from(AutomationFrameworkStopRequest.encode(value).finish()),
        requestDeserialize: (value) => AutomationFrameworkStopRequest.decode(value),
        responseSerialize: (value) => Buffer.from(AutomationFrameworkStopResponse.encode(value).finish()),
        responseDeserialize: (value) => AutomationFrameworkStopResponse.decode(value),
    },
    testOrchestration: {
        path: "/browserstack.sdk.v1.SDK/TestOrchestration",
        requestStream: false,
        responseStream: false,
        requestSerialize: (value) => Buffer.from(TestOrchestrationRequest.encode(value).finish()),
        requestDeserialize: (value) => TestOrchestrationRequest.decode(value),
        responseSerialize: (value) => Buffer.from(TestOrchestrationResponse.encode(value).finish()),
        responseDeserialize: (value) => TestOrchestrationResponse.decode(value),
    },
    findNearestHub: {
        path: "/browserstack.sdk.v1.SDK/FindNearestHub",
        requestStream: false,
        responseStream: false,
        requestSerialize: (value) => Buffer.from(FindNearestHubRequest.encode(value).finish()),
        requestDeserialize: (value) => FindNearestHubRequest.decode(value),
        responseSerialize: (value) => Buffer.from(FindNearestHubResponse.encode(value).finish()),
        responseDeserialize: (value) => FindNearestHubResponse.decode(value),
    },
    aiBrowserExtension: {
        path: "/browserstack.sdk.v1.SDK/AIBrowserExtension",
        requestStream: false,
        responseStream: false,
        requestSerialize: (value) => Buffer.from(AIBrowserExtensionRequest.encode(value).finish()),
        requestDeserialize: (value) => AIBrowserExtensionRequest.decode(value),
        responseSerialize: (value) => Buffer.from(AIBrowserExtensionResponse.encode(value).finish()),
        responseDeserialize: (value) => AIBrowserExtensionResponse.decode(value),
    },
    aiSelfHealStep: {
        path: "/browserstack.sdk.v1.SDK/AISelfHealStep",
        requestStream: false,
        responseStream: false,
        requestSerialize: (value) => Buffer.from(AISelfHealStepRequest.encode(value).finish()),
        requestDeserialize: (value) => AISelfHealStepRequest.decode(value),
        responseSerialize: (value) => Buffer.from(AISelfHealStepResponse.encode(value).finish()),
        responseDeserialize: (value) => AISelfHealStepResponse.decode(value),
    },
    aiSelfHealGetResult: {
        path: "/browserstack.sdk.v1.SDK/AISelfHealGetResult",
        requestStream: false,
        responseStream: false,
        requestSerialize: (value) => Buffer.from(AISelfHealGetRequest.encode(value).finish()),
        requestDeserialize: (value) => AISelfHealGetRequest.decode(value),
        responseSerialize: (value) => Buffer.from(AISelfHealGetResponse.encode(value).finish()),
        responseDeserialize: (value) => AISelfHealGetResponse.decode(value),
    },
    accessibilityConfig: {
        path: "/browserstack.sdk.v1.SDK/AccessibilityConfig",
        requestStream: false,
        responseStream: false,
        requestSerialize: (value) => Buffer.from(AccessibilityConfigRequest.encode(value).finish()),
        requestDeserialize: (value) => AccessibilityConfigRequest.decode(value),
        responseSerialize: (value) => Buffer.from(AccessibilityConfigResponse.encode(value).finish()),
        responseDeserialize: (value) => AccessibilityConfigResponse.decode(value),
    },
    observabilityConfig: {
        path: "/browserstack.sdk.v1.SDK/ObservabilityConfig",
        requestStream: false,
        responseStream: false,
        requestSerialize: (value) => Buffer.from(ObservabilityConfigRequest.encode(value).finish()),
        requestDeserialize: (value) => ObservabilityConfigRequest.decode(value),
        responseSerialize: (value) => Buffer.from(ObservabilityConfigResponse.encode(value).finish()),
        responseDeserialize: (value) => ObservabilityConfigResponse.decode(value),
    },
    accessibilityResult: {
        path: "/browserstack.sdk.v1.SDK/AccessibilityResult",
        requestStream: false,
        responseStream: false,
        requestSerialize: (value) => Buffer.from(AccessibilityResultRequest.encode(value).finish()),
        requestDeserialize: (value) => AccessibilityResultRequest.decode(value),
        responseSerialize: (value) => Buffer.from(AccessibilityResultResponse.encode(value).finish()),
        responseDeserialize: (value) => AccessibilityResultResponse.decode(value),
    },
    testFrameworkEvent: {
        path: "/browserstack.sdk.v1.SDK/TestFrameworkEvent",
        requestStream: false,
        responseStream: false,
        requestSerialize: (value) => Buffer.from(TestFrameworkEventRequest.encode(value).finish()),
        requestDeserialize: (value) => TestFrameworkEventRequest.decode(value),
        responseSerialize: (value) => Buffer.from(TestFrameworkEventResponse.encode(value).finish()),
        responseDeserialize: (value) => TestFrameworkEventResponse.decode(value),
    },
    testSessionEvent: {
        path: "/browserstack.sdk.v1.SDK/TestSessionEvent",
        requestStream: false,
        responseStream: false,
        requestSerialize: (value) => Buffer.from(TestSessionEventRequest.encode(value).finish()),
        requestDeserialize: (value) => TestSessionEventRequest.decode(value),
        responseSerialize: (value) => Buffer.from(TestSessionEventResponse.encode(value).finish()),
        responseDeserialize: (value) => TestSessionEventResponse.decode(value),
    },
    enqueueTestEvent: {
        path: "/browserstack.sdk.v1.SDK/EnqueueTestEvent",
        requestStream: false,
        responseStream: false,
        requestSerialize: (value) => Buffer.from(EnqueueTestEventRequest.encode(value).finish()),
        requestDeserialize: (value) => EnqueueTestEventRequest.decode(value),
        responseSerialize: (value) => Buffer.from(EnqueueTestEventResponse.encode(value).finish()),
        responseDeserialize: (value) => EnqueueTestEventResponse.decode(value),
    },
    logCreatedEvent: {
        path: "/browserstack.sdk.v1.SDK/LogCreatedEvent",
        requestStream: false,
        responseStream: false,
        requestSerialize: (value) => Buffer.from(LogCreatedEventRequest.encode(value).finish()),
        requestDeserialize: (value) => LogCreatedEventRequest.decode(value),
        responseSerialize: (value) => Buffer.from(LogCreatedEventResponse.encode(value).finish()),
        responseDeserialize: (value) => LogCreatedEventResponse.decode(value),
    },
    fetchDriverExecuteParamsEvent: {
        path: "/browserstack.sdk.v1.SDK/FetchDriverExecuteParamsEvent",
        requestStream: false,
        responseStream: false,
        requestSerialize: (value) => Buffer.from(FetchDriverExecuteParamsEventRequest.encode(value).finish()),
        requestDeserialize: (value) => FetchDriverExecuteParamsEventRequest.decode(value),
        responseSerialize: (value) => Buffer.from(FetchDriverExecuteParamsEventResponse.encode(value).finish()),
        responseDeserialize: (value) => FetchDriverExecuteParamsEventResponse.decode(value),
    },
};
export const SDKClient = makeGenericClientConstructor(SDKService, "browserstack.sdk.v1.SDK");
