// appium-uiautomator2-driver/build/test/unit/uiautomator2-specs.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const appium_adb_1 = require("appium-adb");
const sinon_1 = __importDefault(require("sinon"));
const uiautomator2_1 = require("../../lib/uiautomator2");
const helpers = __importStar(require("../../lib/helpers"));
const logger_1 = __importDefault(require("../../lib/logger"));
const chai_1 = require("chai");
describe('UiAutomator2', function () {
    let uiautomator2;
    const adb = new appium_adb_1.ADB();
    const serverApk = {
        'appPath': 'path/to/appium-uiautomator2-server.apk',
        'appId': 'io.appium.uiautomator2.server'
    };
    const serverTestApk = {
        'appPath': 'path/to/appium-uiautomator2-server-test.apk',
        'appId': 'io.appium.uiautomator2.server.test'
    };
    const defaultUIA2ServerOptions = {
        systemPort: 4724,
        host: 'localhost',
        disableWindowAnimation: false
    };
    describe('shouldUninstallServerPackages', function () {
        beforeEach(function () {
            uiautomator2 = new uiautomator2_1.UiAutomator2Server(logger_1.default, {
                adb, ...defaultUIA2ServerOptions
            });
        });
        it('with newer servers are installed', function () {
            // @ts-expect-error - private method
            (0, chai_1.expect)(uiautomator2.shouldUninstallServerPackages([
                {
                    'installState': adb.APP_INSTALL_STATE.NEWER_VERSION_INSTALLED,
                    ...serverApk
                },
                {
                    'installState': adb.APP_INSTALL_STATE.NEWER_VERSION_INSTALLED,
                    ...serverTestApk
                }
            ])).to.be.true;
        }),
            it('with newer server is installed but the other could be old one', function () {
                // Then, enforce to uninstall all apks
                // @ts-expect-error - private method
                (0, chai_1.expect)(uiautomator2.shouldUninstallServerPackages([
                    {
                        'installState': adb.APP_INSTALL_STATE.NEWER_VERSION_INSTALLED,
                        ...serverApk
                    },
                    {
                        'installState': adb.APP_INSTALL_STATE.OLDER_VERSION_INSTALLED,
                        ...serverTestApk
                    }
                ])).to.be.true;
            }),
            it('with newer server is installed', function () {
                // @ts-expect-error - private method
                (0, chai_1.expect)(uiautomator2.shouldUninstallServerPackages([
                    {
                        'installState': adb.APP_INSTALL_STATE.SAME_VERSION_INSTALLED,
                        ...serverApk
                    },
                    {
                        'installState': adb.APP_INSTALL_STATE.SAME_VERSION_INSTALLED,
                        ...serverTestApk
                    }
                ])).to.be.false;
            }),
            it('with older servers are installed', function () {
                // then, installing newer serves are sufficient.
                // @ts-expect-error - private method
                (0, chai_1.expect)(uiautomator2.shouldUninstallServerPackages([
                    {
                        'installState': adb.APP_INSTALL_STATE.OLDER_VERSION_INSTALLED,
                        ...serverApk
                    },
                    {
                        'installState': adb.APP_INSTALL_STATE.OLDER_VERSION_INSTALLED,
                        ...serverTestApk
                    }
                ])).to.be.false;
            }),
            it('with no server are installed', function () {
                // @ts-expect-error - private method
                (0, chai_1.expect)(uiautomator2.shouldUninstallServerPackages([
                    {
                        'installState': adb.APP_INSTALL_STATE.NOT_INSTALLED,
                        ...serverApk
                    },
                    {
                        'installState': adb.APP_INSTALL_STATE.NOT_INSTALLED,
                        ...serverTestApk
                    }
                ])).to.be.false;
            });
    });
    describe('shouldInstallServerPackages', function () {
        beforeEach(function () {
            uiautomator2 = new uiautomator2_1.UiAutomator2Server(logger_1.default, {
                adb, ...defaultUIA2ServerOptions
            });
        });
        it('with newer servers are installed', function () {
            // @ts-expect-error - private method
            (0, chai_1.expect)(uiautomator2.shouldInstallServerPackages([
                {
                    'installState': adb.APP_INSTALL_STATE.NEWER_VERSION_INSTALLED,
                    ...serverApk
                },
                {
                    'installState': adb.APP_INSTALL_STATE.NEWER_VERSION_INSTALLED,
                    ...serverTestApk
                }
                // since installation may fail
            ])).to.be.false;
        }),
            it('with newer server is installed but the other could be old one', function () {
                // Then, enforce to uninstall all apks
                // @ts-expect-error - private method
                (0, chai_1.expect)(uiautomator2.shouldInstallServerPackages([
                    {
                        'installState': adb.APP_INSTALL_STATE.NEWER_VERSION_INSTALLED,
                        ...serverApk
                    },
                    {
                        'installState': adb.APP_INSTALL_STATE.OLDER_VERSION_INSTALLED,
                        ...serverTestApk
                    }
                ])).to.be.true;
            }),
            it('with newer server is installed', function () {
                // @ts-expect-error - private method
                (0, chai_1.expect)(uiautomator2.shouldInstallServerPackages([
                    {
                        'installState': adb.APP_INSTALL_STATE.SAME_VERSION_INSTALLED,
                        ...serverApk
                    },
                    {
                        'installState': adb.APP_INSTALL_STATE.SAME_VERSION_INSTALLED,
                        ...serverTestApk
                    }
                ])).to.be.false;
            }),
            it('with older servers are installed', function () {
                // then, installing newer serves are sufficient.
                // @ts-expect-error - private method
                (0, chai_1.expect)(uiautomator2.shouldInstallServerPackages([
                    {
                        'installState': adb.APP_INSTALL_STATE.OLDER_VERSION_INSTALLED,
                        ...serverApk
                    },
                    {
                        'installState': adb.APP_INSTALL_STATE.OLDER_VERSION_INSTALLED,
                        ...serverTestApk
                    }
                ])).to.be.true;
            }),
            it('with no server are installed', function () {
                // @ts-expect-error - private method
                (0, chai_1.expect)(uiautomator2.shouldInstallServerPackages([
                    {
                        'installState': adb.APP_INSTALL_STATE.NOT_INSTALLED,
                        ...serverApk
                    },
                    {
                        'installState': adb.APP_INSTALL_STATE.NOT_INSTALLED,
                        ...serverTestApk
                    }
                ])).to.be.true;
            });
    });
    describe('installServerApk', function () {
        let mockAdb;
        let mockHelpers;
        let testAdb;
        beforeEach(function () {
            testAdb = new appium_adb_1.ADB();
            mockAdb = sinon_1.default.mock(testAdb);
            mockHelpers = sinon_1.default.mock(helpers);
            uiautomator2 = new uiautomator2_1.UiAutomator2Server(logger_1.default, {
                adb: testAdb, ...defaultUIA2ServerOptions
            });
        });
        afterEach(function () {
            mockAdb.verify();
            mockHelpers.verify();
            mockAdb.restore();
            mockHelpers.restore();
        });
        it('new server and server.test are older than installed version', async function () {
            mockHelpers.expects('isWriteable').never();
            // SERVER_PACKAGE_ID
            mockAdb.expects('getApplicationInstallState').once()
                .returns(adb.APP_INSTALL_STATE.NEWER_VERSION_INSTALLED);
            mockAdb.expects('uninstallApk').twice();
            mockAdb.expects('install').twice();
            mockAdb.expects('isAppInstalled')
                .withExactArgs(uiautomator2_1.SERVER_TEST_PACKAGE_ID)
                .once().returns(true);
            mockAdb.expects('shell')
                .withExactArgs(['pm', 'list', 'instrumentation'])
                .once().returns(uiautomator2_1.INSTRUMENTATION_TARGET);
            await uiautomator2.installServerApk();
        });
        it('new server and server.test are newer than installed version', async function () {
            mockHelpers.expects('isWriteable').never();
            // SERVER_PACKAGE_ID
            mockAdb.expects('getApplicationInstallState').once()
                .returns(adb.APP_INSTALL_STATE.OLDER_VERSION_INSTALLED);
            mockAdb.expects('isAppInstalled')
                .withExactArgs(uiautomator2_1.SERVER_TEST_PACKAGE_ID)
                .once().returns(true);
            mockAdb.expects('uninstallApk').never();
            mockAdb.expects('install').twice();
            mockAdb.expects('shell')
                .withExactArgs(['pm', 'list', 'instrumentation'])
                .once().returns(uiautomator2_1.INSTRUMENTATION_TARGET);
            await uiautomator2.installServerApk();
        });
        it('new server and server.test are the same as installed version', async function () {
            mockHelpers.expects('isWriteable').never();
            // SERVER_PACKAGE_ID
            mockAdb.expects('getApplicationInstallState').once()
                .returns(adb.APP_INSTALL_STATE.SAME_VERSION_INSTALLED);
            mockAdb.expects('isAppInstalled')
                .withExactArgs(uiautomator2_1.SERVER_TEST_PACKAGE_ID)
                .once().returns(true);
            mockAdb.expects('uninstallApk').never();
            mockAdb.expects('install').never();
            mockAdb.expects('shell')
                .withExactArgs(['pm', 'list', 'instrumentation'])
                .once().returns(uiautomator2_1.INSTRUMENTATION_TARGET);
            await uiautomator2.installServerApk();
        });
        it('new server and server.test are not installed', async function () {
            mockHelpers.expects('isWriteable').never();
            // SERVER_PACKAGE_ID
            mockAdb.expects('getApplicationInstallState').once()
                .returns(adb.APP_INSTALL_STATE.NOT_INSTALLED);
            // SERVER_TEST_PACKAGE_ID
            mockAdb.expects('isAppInstalled')
                .withExactArgs(uiautomator2_1.SERVER_TEST_PACKAGE_ID)
                .once().returns(false);
            mockAdb.expects('uninstallApk').never();
            mockAdb.expects('install').twice();
            mockAdb.expects('shell').withExactArgs(['pm', 'list', 'instrumentation'])
                .once().returns(uiautomator2_1.INSTRUMENTATION_TARGET);
            await uiautomator2.installServerApk();
        });
        it('version numbers of new server and server.test are unknown', async function () {
            mockHelpers.expects('isWriteable').never();
            // SERVER_PACKAGE_ID
            mockAdb.expects('getApplicationInstallState').once()
                .returns(adb.APP_INSTALL_STATE.UNKNOWN);
            mockAdb.expects('isAppInstalled')
                .withExactArgs(uiautomator2_1.SERVER_TEST_PACKAGE_ID)
                .once().returns(false);
            mockAdb.expects('uninstallApk').twice();
            mockAdb.expects('install').twice();
            mockAdb.expects('shell')
                .withExactArgs(['pm', 'list', 'instrumentation'])
                .once().returns(uiautomator2_1.INSTRUMENTATION_TARGET);
            await uiautomator2.installServerApk();
        });
        it('a server is installed but server.test is not', async function () {
            mockHelpers.expects('isWriteable').never();
            // SERVER_PACKAGE_ID
            mockAdb.expects('getApplicationInstallState').once()
                .returns(adb.APP_INSTALL_STATE.SAME_VERSION_INSTALLED);
            mockAdb.expects('isAppInstalled')
                .withExactArgs(uiautomator2_1.SERVER_TEST_PACKAGE_ID)
                .once().returns(false);
            mockAdb.expects('uninstallApk').twice();
            mockAdb.expects('install').twice();
            mockAdb.expects('shell')
                .withExactArgs(['pm', 'list', 'instrumentation'])
                .once().returns(uiautomator2_1.INSTRUMENTATION_TARGET);
            await uiautomator2.installServerApk();
        });
    });
});
