// @browserstack/wdio-browserstack-service/dist/index.js
// gRPC Service Client and credentials
export { SDKClient } from './generated/sdk.js';
export { credentials as grpcCredentials, Channel as grpcChannel } from '@grpc/grpc-js';
// Create factory objects with create methods for backward compatibility
export const StartBinSessionRequestConstructor = {
    create: (init = {}) => {
        return {
            binSessionId: '',
            sdkLanguage: '',
            sdkVersion: '',
            pathProject: '',
            pathConfig: '',
            envVars: {},
            cliArgs: [],
            frameworks: [],
            frameworkVersions: {},
            language: '',
            languageVersion: '',
            testFramework: '',
            wdioConfig: '',
            ...init
        };
    }
};
export const StopBinSessionRequestConstructor = {
    create: (init = {}) => {
        return {
            binSessionId: '',
            exitCode: 0,
            exitSignal: '',
            exitReason: '',
            customMetadata: '',
            ...init
        };
    }
};
export const ConnectBinSessionRequestConstructor = {
    create: (init = {}) => {
        return {
            binSessionId: '',
            platformIndex: 0,
            ...init
        };
    }
};
export const TestFrameworkEventRequestConstructor = {
    create: (init = {}) => {
        return {
            binSessionId: '',
            platformIndex: 0,
            testFrameworkName: '',
            testFrameworkVersion: '',
            uuid: '',
            eventJson: new Uint8Array(),
            testFrameworkState: '',
            testHookState: '',
            startedAt: '',
            endedAt: '',
            executionContext: undefined,
            ...init
        };
    }
};
export const TestSessionEventRequestConstructor = {
    create: (init = {}) => {
        return {
            binSessionId: '',
            platformIndex: 0,
            testFrameworkName: '',
            testFrameworkVersion: '',
            testFrameworkState: '',
            testHookState: '',
            testUuid: '',
            automationSessions: [],
            executionContext: undefined,
            capabilities: new Uint8Array(),
            ...init
        };
    }
};
export const ExecutionContextConstructor = {
    create: (init = {}) => {
        return {
            threadId: '',
            processId: '',
            hash: '',
            ...init
        };
    }
};
export const LogCreatedEventRequestConstructor = {
    create: (init = {}) => {
        return {
            binSessionId: '',
            platformIndex: 0,
            logs: [],
            executionContext: undefined,
            ...init
        };
    }
};
export const LogCreatedEventRequest_LogEntryConstructor = {
    create: (init = {}) => {
        return {
            testFrameworkName: '',
            testFrameworkVersion: '',
            testFrameworkState: '',
            timestamp: '',
            uuid: '',
            kind: '',
            message: new Uint8Array(),
            level: '',
            fileName: '',
            fileSize: 0,
            filePath: '',
            ...init
        };
    }
};
export const AutomationSessionConstructor = {
    create: (init = {}) => {
        return {
            provider: '',
            frameworkName: '',
            frameworkVersion: '',
            frameworkSessionId: '',
            ref: '',
            hubUrl: '',
            ...init
        };
    }
};
export const DriverInitRequestConstructor = {
    create: (init = {}) => {
        return {
            binSessionId: '',
            platformIndex: 0,
            ref: '',
            userInputParams: new Uint8Array(),
            ...init
        };
    }
};
export const FetchDriverExecuteParamsEventRequestConstructor = {
    create: (init = {}) => {
        return {
            binSessionId: '',
            product: '',
            scriptName: '',
            ...init
        };
    }
};
